// ravendb/dist/Documents/Operations/Refresh/ConfigureRefreshOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigureRefreshOperation = void 0;
const RavenCommand_1 = require("../../../Http/RavenCommand");
const RaftIdGenerator_1 = require("../../../Utility/RaftIdGenerator");
const Exceptions_1 = require("../../../Exceptions");
class ConfigureRefreshOperation {
    constructor(configuration) {
        if (!configuration) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Configuration cannot be null");
        }
        this._configuration = configuration;
    }
    getCommand(conventions) {
        return new ConfigureRefreshCommand(this._configuration);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.ConfigureRefreshOperation = ConfigureRefreshOperation;
class ConfigureRefreshCommand extends RavenCommand_1.RavenCommand {
    constructor(configuration) {
        super();
        this._configuration = configuration;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/refresh/config";
        const body = this._serializer.serialize(this._configuration);
        return {
            uri,
            method: "POST",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_1.RaftIdGenerator.newId();
    }
}
