// ravendb/dist/Documents/Operations/GetDetailedCollectionStatisticsOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetDetailedCollectionStatisticsOperation = void 0;
const RavenCommand_1 = require("../../Http/RavenCommand");
class GetDetailedCollectionStatisticsOperation {
    getCommand(conventions) {
        return new GetDetailedCollectionStatisticsCommand();
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.GetDetailedCollectionStatisticsOperation = GetDetailedCollectionStatisticsOperation;
class GetDetailedCollectionStatisticsCommand extends RavenCommand_1.RavenCommand {
    get isReadRequest() {
        return true;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/collections/stats/detailed";
        return {
            method: "GET",
            uri
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
}
