// ravendb/dist/Documents/Operations/GetOngoingTaskInfoOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetOngoingTaskInfoOperation = void 0;
const TypeUtil_1 = require("../../Utility/TypeUtil");
const RavenCommand_1 = require("../../Http/RavenCommand");
const index_1 = require("../../Exceptions/index");
const RavenEtlConfiguration_1 = require("./Etl/RavenEtlConfiguration");
const SqlEtlConfiguration_1 = require("./Etl/Sql/SqlEtlConfiguration");
class GetOngoingTaskInfoOperation {
    constructor(taskIdOrName, type) {
        if (TypeUtil_1.TypeUtil.isString(taskIdOrName)) {
            this._taskName = taskIdOrName;
        }
        else {
            this._taskId = taskIdOrName;
        }
        this._type = type;
        if (type === "PullReplicationAsHub") {
            (0, index_1.throwError)("InvalidArgumentException", "PullReplicationAsHub type is not supported. " +
                "Please use GetPullReplicationTasksInfoOperation instead.");
        }
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new GetOngoingTaskInfoCommand(this._taskName || this._taskId, this._type, conventions);
    }
}
exports.GetOngoingTaskInfoOperation = GetOngoingTaskInfoOperation;
class GetOngoingTaskInfoCommand extends RavenCommand_1.RavenCommand {
    constructor(taskIdOrName, type, documentConventions) {
        super();
        if (TypeUtil_1.TypeUtil.isString(taskIdOrName)) {
            this._taskName = taskIdOrName;
        }
        else {
            this._taskId = taskIdOrName;
        }
        this._type = type;
        this._conventions = documentConventions;
    }
    createRequest(node) {
        const uri = this._taskName
            ? node.url + "/databases/" + node.database + "/task?taskName=" + encodeURIComponent(this._taskName) + "&type=" + this._type
            : node.url + "/databases/" + node.database + "/task?key=" + this._taskId + "&type=" + this._type;
        return {
            uri,
            method: "GET"
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        let body = null;
        const results = await this._defaultPipeline(_ => body = _)
            .process(bodyStream);
        let nestedTypes = {};
        switch (this._type) {
            case "Replication":
                break;
            case "RavenEtl":
                nestedTypes = {
                    configuration: "RavenEtlConfiguration"
                };
                break;
            case "SqlEtl":
                nestedTypes = {
                    configuration: "SqlEtlConfiguration"
                };
                break;
            case "Subscription":
                nestedTypes = {
                    lastBatchAckTime: "date",
                    lastClientConnectionTime: "date"
                };
                break;
            case "PullReplicationAsSink":
                break;
            case "Backup":
                nestedTypes = {
                    lastFullBackup: "date",
                    lastIncrementalBackup: "date",
                    "onGoingBackup.startTime": "date",
                    "nextBackup.dateTime": "date"
                };
                break;
        }
        this.result = this._reviveResultTypes(results, this._conventions, {
            nestedTypes
        }, knownTypes);
        return body;
    }
    get isReadRequest() {
        return false;
    }
}
const knownTypes = new Map([
    [RavenEtlConfiguration_1.RavenEtlConfiguration.name, RavenEtlConfiguration_1.RavenEtlConfiguration],
    [SqlEtlConfiguration_1.SqlEtlConfiguration.name, SqlEtlConfiguration_1.SqlEtlConfiguration]
]);
