// ravendb/dist/Documents/Operations/OperationAbstractions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OperationExceptionResult = exports.AwaitableOperation = exports.AwaitableMaintenanceOperation = exports.AwaitableServerOperation = exports.AbstractAwaitableOperation = void 0;
class AbstractAwaitableOperation {
    get resultType() {
        return "OperationId";
    }
}
exports.AbstractAwaitableOperation = AbstractAwaitableOperation;
class AwaitableServerOperation extends AbstractAwaitableOperation {
    getCommand(conventions) {
        throw new Error("getCommand() must be implemented in extending class.");
    }
}
exports.AwaitableServerOperation = AwaitableServerOperation;
class AwaitableMaintenanceOperation extends AbstractAwaitableOperation {
    getCommand(conventions) {
        throw new Error("getCommand() must be implemented in extending class.");
    }
}
exports.AwaitableMaintenanceOperation = AwaitableMaintenanceOperation;
class AwaitableOperation extends AbstractAwaitableOperation {
    getCommand(store, conventions, httpCache) {
        throw new Error("getCommand() must be implemented in extending class.");
    }
}
exports.AwaitableOperation = AwaitableOperation;
class OperationExceptionResult {
}
exports.OperationExceptionResult = OperationExceptionResult;
