// ravendb/dist/Documents/Commands/Batches/DeleteAttachmentCommandData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteAttachmentCommandData = void 0;
const StringUtil_1 = require("../../../Utility/StringUtil");
const Exceptions_1 = require("../../../Exceptions");
class DeleteAttachmentCommandData {
    constructor(documentId, name, changeVector) {
        this.type = "AttachmentDELETE";
        if (StringUtil_1.StringUtil.isNullOrWhitespace(documentId)) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "DocumentId cannot be null");
        }
        if (StringUtil_1.StringUtil.isNullOrWhitespace(name)) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Name cannot be null");
        }
        this.id = documentId;
        this.name = name;
        this.changeVector = changeVector;
    }
    serialize(conventions) {
        return {
            Id: this.id,
            Name: this.name,
            ChangeVector: this.changeVector,
            Type: "AttachmentDELETE"
        };
    }
}
exports.DeleteAttachmentCommandData = DeleteAttachmentCommandData;
