// ravendb/dist/Documents/Commands/Batches/TimeSeriesBatchCommandData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeSeriesBatchCommandData = void 0;
const TimeSeriesOperation_1 = require("../../Operations/TimeSeries/TimeSeriesOperation");
const Exceptions_1 = require("../../../Exceptions");
class TimeSeriesBatchCommandData {
    constructor(documentId, name, appends, deletes) {
        if (!documentId) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "DocumentId cannot be null");
        }
        if (!name) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Name cannot be null");
        }
        this._id = documentId;
        this._name = name;
        this._timeSeries = new TimeSeriesOperation_1.TimeSeriesOperation();
        this._timeSeries.name = name;
        if (appends) {
            for (const appendOperation of appends) {
                this._timeSeries.append(appendOperation);
            }
        }
        if (deletes) {
            for (const deleteOperation of deletes) {
                this._timeSeries.delete(deleteOperation);
            }
        }
    }
    get id() {
        return this._id;
    }
    set id(value) {
        this._id = value;
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
    }
    get changeVector() {
        return null;
    }
    get type() {
        return "TimeSeries";
    }
    get timeSeries() {
        return this._timeSeries;
    }
    serialize(conventions) {
        return {
            Id: this._id,
            TimeSeries: this._timeSeries.serialize(conventions),
            Type: "TimeSeries"
        };
    }
    onBeforeSaveChanges(session) {
    }
}
exports.TimeSeriesBatchCommandData = TimeSeriesBatchCommandData;
