// ravendb/dist/Documents/Operations/Analyzers/PutAnalyzersOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PutAnalyzersOperation = void 0;
const Exceptions_1 = require("../../../Exceptions");
const RavenCommand_1 = require("../../../Http/RavenCommand");
const RaftIdGenerator_1 = require("../../../Utility/RaftIdGenerator");
const HttpUtil_1 = require("../../../Utility/HttpUtil");
class PutAnalyzersOperation {
    constructor(...analyzersToAdd) {
        if (!analyzersToAdd || analyzersToAdd.length === 0) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "AnalyzersToAdd cannot be null or empty");
        }
        this._analyzersToAdd = analyzersToAdd;
    }
    getCommand(conventions) {
        return new PutAnalyzersCommand(conventions, this._analyzersToAdd);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.PutAnalyzersOperation = PutAnalyzersOperation;
class PutAnalyzersCommand extends RavenCommand_1.RavenCommand {
    constructor(conventions, analyzersToAdd) {
        super();
        if (!conventions) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Conventions cannot be null");
        }
        if (!analyzersToAdd) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Analyzers to add cannot be null");
        }
        this._analyzersToAdd = [];
        for (const analyzerDefinition of analyzersToAdd) {
            if (!analyzerDefinition.name) {
                (0, Exceptions_1.throwError)("InvalidArgumentException", "Name cannot be null");
            }
            this._analyzersToAdd.push(conventions.objectMapper.toObjectLiteral(analyzerDefinition));
        }
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/analyzers";
        const headers = HttpUtil_1.HeadersBuilder
            .create()
            .typeAppJson()
            .build();
        const body = this._serializer
            .serialize({
            Analyzers: this._analyzersToAdd
        });
        return {
            uri,
            method: "PUT",
            headers,
            body
        };
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_1.RaftIdGenerator.newId();
    }
}
