// ravendb/dist/Documents/Operations/OngoingTasks/GetPullReplicationHubTasksInfoOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetPullReplicationHubTasksInfoOperation = void 0;
const RavenCommand_1 = require("../../../Http/RavenCommand");
class GetPullReplicationHubTasksInfoOperation {
    constructor(taskId) {
        this._taskId = taskId;
    }
    getCommand(conventions) {
        return new GetPullReplicationTasksInfoCommand(this._taskId);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.GetPullReplicationHubTasksInfoOperation = GetPullReplicationHubTasksInfoOperation;
class GetPullReplicationTasksInfoCommand extends RavenCommand_1.RavenCommand {
    constructor(taskId) {
        super();
        this._taskId = taskId;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/tasks/pull-replication/hub?key=" + this._taskId;
        return {
            method: "GET",
            uri
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        return this._parseResponseDefaultAsync(bodyStream);
    }
    get isReadRequest() {
        return false;
    }
}
