// ravendb/dist/Http/RavenCommandResponsePipeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RavenCommandResponsePipeline = void 0;
const events_1 = require("events");
const Parser = require("stream-json/Parser");
const ObjectKeyCaseTransformStream_1 = require("../Mapping/Json/Streams/ObjectKeyCaseTransformStream");
const Conventions_1 = require("../Mapping/Json/Conventions");
const StreamUtil = require("../Utility/StreamUtil");
const stream = require("readable-stream");
const CollectResultStream_1 = require("../Mapping/Json/Streams/CollectResultStream");
const Exceptions_1 = require("../Exceptions");
const TypeUtil_1 = require("../Utility/TypeUtil");
const Asm = require("stream-json/Assembler");
const StringBuilder_1 = require("../Utility/StringBuilder");
const Parser_1 = require("stream-json/jsonl/Parser");
class RavenCommandResponsePipeline extends events_1.EventEmitter {
    constructor() {
        super();
        this._body = new StringBuilder_1.StringBuilder();
        this._opts = {};
    }
    static create() {
        return new RavenCommandResponsePipeline();
    }
    parseJsonAsync(filters) {
        this._opts.jsonAsync = { filters };
        return this;
    }
    parseJsonSync() {
        this._opts.jsonSync = true;
        return this;
    }
    parseJsonlAsync(valueExtractor, options = {}) {
        const transforms = options?.transforms ?? [];
        const extractItemTransform = new stream.Transform({
            objectMode: true,
            transform(chunk, encoding, callback) {
                const value = valueExtractor(chunk["value"]);
                if (!value) {
                    return callback();
                }
                callback(null, { ...chunk, value });
            }
        });
        transforms.unshift(extractItemTransform);
        this._opts.jsonlAsync = {
            transforms
        };
        return this;
    }
    collectBody(callback) {
        this._opts.collectBody = callback || true;
        return this;
    }
    objectKeysTransform(optsOrTransform, profile) {
        if (!this._opts.jsonAsync && !this._opts.jsonSync) {
            (0, Exceptions_1.throwError)("InvalidOperationException", "Cannot use key case transform without doing parseJson() or parseJsonAsync() first.");
        }
        if (!optsOrTransform || typeof optsOrTransform === "string") {
            this._opts.streamKeyCaseTransform =
                (0, Conventions_1.getObjectKeyCaseTransformProfile)(optsOrTransform, profile);
        }
        else {
            this._opts.streamKeyCaseTransform = optsOrTransform;
        }
        if (this._opts.jsonAsync) {
            this._opts.streamKeyCaseTransform.handleKeyValue = true;
        }
        return this;
    }
    collectResult(optsOrReduce, init) {
        if (typeof optsOrReduce === "function") {
            this._opts.collectResult = { reduceResults: optsOrReduce, initResult: init };
        }
        else {
            this._opts.collectResult = optsOrReduce;
        }
        return this;
    }
    stream(src, dst, callback) {
        const streams = this._buildUp(src);
        if (dst) {
            streams.push(dst);
        }
        return stream.pipeline(...streams, TypeUtil_1.TypeUtil.NOOP);
    }
    _appendBody(s) {
        this._body.append(s.toString());
    }
    _buildUp(src) {
        if (!src) {
            (0, Exceptions_1.throwError)("MappingError", "Body stream cannot be null.");
        }
        const opts = this._opts;
        const streams = [src];
        if (opts.collectBody) {
            src.on("data", (chunk) => this._appendBody(chunk));
        }
        if (opts.jsonAsync) {
            const parser = new Parser({ streamValues: false });
            streams.push(parser);
            if (opts.jsonAsync.filters && opts.jsonAsync.filters.length) {
                streams.push(...opts.jsonAsync.filters);
            }
        }
        else if (opts.jsonlAsync) {
            streams.push((0, Parser_1.parser)());
            if (opts.jsonlAsync.transforms) {
                streams.push(...opts.jsonlAsync.transforms);
            }
        }
        else if (opts.jsonSync) {
            const bytesChunks = [];
            const parseJsonSyncTransform = new stream.Transform({
                readableObjectMode: true,
                transform(chunk, enc, callback) {
                    bytesChunks.push(chunk);
                    callback();
                },
                flush(callback) {
                    let str = null;
                    try {
                        str = Buffer.concat(bytesChunks).toString('utf-8');
                    }
                    catch (err) {
                        callback((0, Exceptions_1.getError)("InvalidDataException", `Failed to concat / decode server response`, err));
                        return;
                    }
                    try {
                        callback(null, JSON.parse(str));
                    }
                    catch (err) {
                        callback((0, Exceptions_1.getError)("InvalidOperationException", `Error parsing response: '${str}'.`, err));
                    }
                }
            });
            streams.push(parseJsonSyncTransform);
        }
        if (opts.streamKeyCaseTransform) {
            const handlePath = !!opts.jsonAsync;
            const keyCaseOpts = Object.assign({}, opts.streamKeyCaseTransform, { handlePath });
            streams.push(new ObjectKeyCaseTransformStream_1.ObjectKeyCaseTransformStream(keyCaseOpts));
        }
        return streams;
    }
    process(src) {
        const streams = this._buildUp(src);
        const opts = this._opts;
        let resultPromise;
        if (opts.jsonAsync) {
            const asm = Asm.connectTo(streams[streams.length - 1]);
            resultPromise = new Promise(resolve => {
                asm.on("done", asm => resolve(asm.current));
            });
        }
        else {
            const collectResultOpts = !opts.collectResult || !opts.collectResult.reduceResults
                ? { reduceResults: CollectResultStream_1.lastChunk } : opts.collectResult;
            const collectResult = new CollectResultStream_1.CollectResultStream(collectResultOpts);
            streams.push(collectResult);
            resultPromise = collectResult.promise;
        }
        if (opts.collectBody) {
            resultPromise
                .then(() => {
                const body = this._body.toString();
                this.emit("body", body);
                if (typeof opts.collectBody === "function") {
                    opts.collectBody(body);
                }
            });
        }
        return StreamUtil.pipelineAsync(...streams)
            .then(() => resultPromise);
    }
}
exports.RavenCommandResponsePipeline = RavenCommandResponsePipeline;
