// ravendb/dist/ServerWide/Operations/AddDatabaseNodeOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddDatabaseNodeOperation = void 0;
const Exceptions_1 = require("../../Exceptions");
const RavenCommand_1 = require("../../Http/RavenCommand");
const RaftIdGenerator_1 = require("../../Utility/RaftIdGenerator");
class AddDatabaseNodeOperation {
    constructor(databaseName, node) {
        this._databaseName = databaseName;
        this._node = node;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new AddDatabaseNodeCommand(this._databaseName, this._node);
    }
}
exports.AddDatabaseNodeOperation = AddDatabaseNodeOperation;
class AddDatabaseNodeCommand extends RavenCommand_1.RavenCommand {
    constructor(databaseName, node) {
        super();
        if (!databaseName) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "DatabaseName cannot be null");
        }
        this._databaseName = databaseName;
        this._node = node;
    }
    createRequest(node) {
        let uri = node.url + "/admin/databases/node?name=" + this._databaseName;
        if (this._node) {
            uri += "&node=" + this._node;
        }
        return {
            uri,
            method: "PUT"
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_1.RaftIdGenerator.newId();
    }
}
