// ravendb/dist/ServerWide/Operations/ReorderDatabaseMembersOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReorderDatabaseMembersOperation = void 0;
const Exceptions_1 = require("../../Exceptions");
const RavenCommand_1 = require("../../Http/RavenCommand");
const RaftIdGenerator_1 = require("../../Utility/RaftIdGenerator");
class ReorderDatabaseMembersOperation {
    constructor(database, order, fixed = false) {
        if (!order || order.length === 0) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Order list must contain values");
        }
        this._database = database;
        this._parameters = {
            membersOrder: order,
            fixed
        };
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new ReorderDatabaseMembersCommand(this._database, this._parameters);
    }
}
exports.ReorderDatabaseMembersOperation = ReorderDatabaseMembersOperation;
class ReorderDatabaseMembersCommand extends RavenCommand_1.RavenCommand {
    constructor(databaseName, parameters) {
        super();
        if (!databaseName) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Database cannot be empty");
        }
        this._databaseName = databaseName;
        this._parameters = parameters;
    }
    createRequest(node) {
        const uri = node.url + "/admin/databases/reorder?name=" + this._databaseName;
        const body = this._serializer.serialize(this._parameters);
        return {
            uri,
            method: "POST",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_1.RaftIdGenerator.newId();
    }
}
