// ravendb/dist/ServerWide/Operations/UpdateUnusedDatabasesOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateUnusedDatabasesOperation = void 0;
const StringUtil_1 = require("../../Utility/StringUtil");
const index_1 = require("../../Exceptions/index");
const RavenCommand_1 = require("../../Http/RavenCommand");
const RaftIdGenerator_1 = require("../../Utility/RaftIdGenerator");
class UpdateUnusedDatabasesOperation {
    constructor(database, unusedDatabaseIds) {
        if (StringUtil_1.StringUtil.isNullOrEmpty(database)) {
            (0, index_1.throwError)("InvalidArgumentException", "Database cannot be null");
        }
        this._database = database;
        this._parameters = {
            databaseIds: unusedDatabaseIds
        };
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new UpdateUnusedDatabasesCommand(this._database, this._parameters);
    }
}
exports.UpdateUnusedDatabasesOperation = UpdateUnusedDatabasesOperation;
class UpdateUnusedDatabasesCommand extends RavenCommand_1.RavenCommand {
    constructor(database, parameters) {
        super();
        this._database = database;
        this._parameters = parameters;
    }
    createRequest(node) {
        const uri = node.url + "/admin/databases/unused-ids?name=" + this._database;
        const body = this._serializer.serialize(this._parameters);
        return {
            uri,
            method: "POST",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_1.RaftIdGenerator.newId();
    }
}
