// ravendb/dist/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeSeriesCollectionConfiguration = void 0;
const TimeSeriesPolicy_1 = require("./TimeSeriesPolicy");
const RawTimeSeriesPolicy_1 = require("./RawTimeSeriesPolicy");
class TimeSeriesCollectionConfiguration {
    constructor() {
        this.policies = [];
        this.rawPolicy = RawTimeSeriesPolicy_1.RawTimeSeriesPolicy.DEFAULT_POLICY;
    }
    static isRaw(policy) {
        return RawTimeSeriesPolicy_1.RawTimeSeriesPolicy.DEFAULT_POLICY.name === policy.name;
    }
    serialize() {
        return {
            Disabled: this.disabled,
            Policies: this.policies ? this.policies.map(p => p.serialize()) : null,
            RawPolicy: this.rawPolicy ? this.rawPolicy.serialize() : null
        };
    }
    static parse(collectionRaw) {
        const configuration = new TimeSeriesCollectionConfiguration();
        configuration.disabled = collectionRaw.Disabled;
        configuration.policies = collectionRaw.Policies.map(x => TimeSeriesPolicy_1.TimeSeriesPolicy.parse(x));
        configuration.rawPolicy = RawTimeSeriesPolicy_1.RawTimeSeriesPolicy.parse(collectionRaw.RawPolicy);
        return configuration;
    }
}
exports.TimeSeriesCollectionConfiguration = TimeSeriesCollectionConfiguration;
