// ravendb/dist/Documents/Operations/TimeSeries/TimeSeriesOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteOperation = exports.AppendOperation = exports.TimeSeriesOperation = void 0;
const DateUtil_1 = require("../../../Utility/DateUtil");
class TimeSeriesOperation {
    constructor(name) {
        this.name = name;
    }
    serialize(conventions) {
        const sortedAppends = this._appends ?
            Array.from(this._appends.values())
                .sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime())
                .map(x => x.serialize(conventions))
            : null;
        return {
            Name: this.name,
            Appends: sortedAppends,
            Deletes: this._deletes ? this._deletes.map(x => x.serialize(conventions)) : null
        };
    }
    append(appendOperation) {
        if (!this._appends) {
            this._appends = new Map();
        }
        const time = appendOperation.timestamp.getTime();
        if (this._appends.has(time)) {
            this._appends.delete(time);
        }
        this._appends.set(time, appendOperation);
    }
    delete(deleteOperation) {
        if (!this._deletes) {
            this._deletes = [];
        }
        this._deletes.push(deleteOperation);
    }
}
exports.TimeSeriesOperation = TimeSeriesOperation;
class AppendOperation {
    constructor(timestamp, values, tag) {
        this.timestamp = timestamp;
        this.values = values;
        this.tag = tag;
    }
    serialize(conventions) {
        return {
            Timestamp: DateUtil_1.DateUtil.utc.stringify(this.timestamp),
            Values: this.values,
            Tag: this.tag || undefined
        };
    }
}
exports.AppendOperation = AppendOperation;
class DeleteOperation {
    constructor(from, to) {
        this.from = from;
        this.to = to;
    }
    serialize(conventions) {
        return {
            From: this.from ? DateUtil_1.DateUtil.utc.stringify(this.from) : null,
            To: this.to ? DateUtil_1.DateUtil.utc.stringify(this.to) : null
        };
    }
}
exports.DeleteOperation = DeleteOperation;
