// ravendb/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetIndexesOperation = exports.GetIndexStatisticsOperation = exports.GetIndexesStatisticsOperation = exports.GetIndexingStatusOperation = exports.EnableIndexOperation = exports.DisableIndexOperation = exports.DeleteIndexErrorsOperation = exports.GetIndexNamesOperation = exports.PutDocumentCommand = exports.PutClientConfigurationOperation = exports.GetClientConfigurationOperation = exports.OperationCompletionAwaiter = exports.MaintenanceOperationExecutor = exports.GetOperationStateOperation = exports.GetStatisticsCommand = exports.GetStatisticsOperation = exports.GetIdentitiesOperation = exports.ExplainQueryCommand = exports.SeedIdentityForCommand = exports.NextIdentityForCommand = exports.DeleteDocumentCommand = exports.KillOperationCommand = exports.GetNextOperationIdCommand = exports.GetCollectionStatisticsOperation = exports.DeleteByQueryOperation = exports.CompareExchangeValueJsonConverter = exports.CompareExchangeSessionValue = exports.DeleteCompareExchangeValueOperation = exports.GetCompareExchangeValuesOperation = exports.CompareExchangeValueResultParser = exports.CompareExchangeValue = exports.CompareExchangeResult = exports.GetCompareExchangeValueOperation = exports.PutCompareExchangeValueOperation = exports.PatchByQueryOperation = exports.PutSortersOperation = exports.DeleteSorterOperation = exports.PatchOperation = exports.PutConnectionStringOperation = exports.CompactDatabaseOperation = exports.GetDetailedCollectionStatisticsOperation = exports.DatabaseHealthCheckOperation = exports.BulkInsertOperation = exports.CreateDatabaseOperation = exports.AddClusterNodeCommand = exports.GetTcpInfoCommand = exports.GetClusterTopologyCommand = exports.GetDatabaseTopologyCommand = exports.BulkInsertConventions = exports.DocumentConventions = void 0;
exports.Json = exports.CONSTANTS = exports.Mapping = exports.TypesAwareObjectMapper = exports.TimeSeriesRangeResult = exports.TimeSeriesPolicy = exports.TimeSeriesOperation = exports.TimeSeriesDetails = exports.TimeSeriesConfiguration = exports.TimeSeriesCollectionConfiguration = exports.TimeSeriesBatchOperation = exports.RemoveTimeSeriesPolicyOperation = exports.RawTimeSeriesPolicy = exports.GetTimeSeriesStatisticsOperation = exports.GetTimeSeriesOperation = exports.GetMultipleTimeSeriesOperation = exports.ConfigureTimeSeriesValueNamesOperation = exports.ConfigureTimeSeriesPolicyOperation = exports.ConfigureTimeSeriesOperation = exports.ConfigureRawTimeSeriesPolicyOperation = exports.TIME_SERIES_ROLLUP_SEPARATOR = exports.DocumentCountersOperation = exports.CounterOperation = exports.CounterBatchOperation = exports.GetCountersOperation = exports.CounterBatch = exports.GetIndexErrorsOperation = exports.GetIndexOperation = exports.SetIndexesPriorityOperation = exports.SetIndexesLockOperation = exports.GetConflictsCommand = exports.UpdatePullReplicationAsSinkOperation = exports.PutPullReplicationAsHubOperation = exports.UpdateExternalReplicationOperation = exports.SeedIdentityForOperation = exports.NextIdentityForOperation = exports.DeleteIndexOperation = exports.ResetIndexOperation = exports.StartIndexOperation = exports.StopIndexOperation = exports.StartIndexingOperation = exports.StopIndexingOperation = exports.PutIndexesOperation = exports.IndexHasChangedOperation = exports.GetTermsOperation = void 0;
require("./Utility/Polyfills");
require("./Types/readable-stream");
var DocumentConventions_1 = require("./Documents/Conventions/DocumentConventions");
Object.defineProperty(exports, "DocumentConventions", { enumerable: true, get: function () { return DocumentConventions_1.DocumentConventions; } });
var BulkInsertConventions_1 = require("./Documents/Conventions/BulkInsertConventions");
Object.defineProperty(exports, "BulkInsertConventions", { enumerable: true, get: function () { return BulkInsertConventions_1.BulkInsertConventions; } });
__exportStar(require("./Types"), exports);
__exportStar(require("./Http/AggressiveCacheOptions"), exports);
__exportStar(require("./Http/ClusterRequestExecutor"), exports);
__exportStar(require("./Http/ClusterTopology"), exports);
__exportStar(require("./Http/CurrentIndexAndNode"), exports);
__exportStar(require("./Http/CurrentIndexAndNodeAndEtag"), exports);
__exportStar(require("./Http/IBroadcast"), exports);
__exportStar(require("./Http/IRaftCommand"), exports);
__exportStar(require("./Http/NodeSelector"), exports);
__exportStar(require("./Http/LoadBalanceBehavior"), exports);
__exportStar(require("./Http/RavenCommand"), exports);
__exportStar(require("./Http/ReadBalanceBehavior"), exports);
__exportStar(require("./Http/RequestExecutor"), exports);
__exportStar(require("./Http/ServerNode"), exports);
__exportStar(require("./Http/StatusCode"), exports);
__exportStar(require("./Http/Topology"), exports);
__exportStar(require("./Http/UriUtility"), exports);
__exportStar(require("./Http/UpdateTopologyParameters"), exports);
__exportStar(require("./ServerWide"), exports);
__exportStar(require("./ServerWide/CompactSettings"), exports);
__exportStar(require("./Documents/Operations/Etl/ConnectionString"), exports);
__exportStar(require("./ServerWide/ModifyOnGoingTaskResult"), exports);
__exportStar(require("./ServerWide/DocumentsCompressionConfiguration"), exports);
__exportStar(require("./ServerWide/DeletionInProgressStatus"), exports);
__exportStar(require("./ServerWide/IDatabaseTaskStatus"), exports);
__exportStar(require("./ServerWide/Operations/BuildNumber"), exports);
__exportStar(require("./ServerWide/Operations/GetBuildNumberOperation"), exports);
__exportStar(require("./ServerWide/Operations/ReorderDatabaseMembersOperation"), exports);
__exportStar(require("./ServerWide/Operations/ConfigureRevisionsForConflictsOperation"), exports);
__exportStar(require("./ServerWide/Operations/UpdateDatabaseOperation"), exports);
__exportStar(require("./ServerWide/Operations/Configuration/GetServerWideBackupConfigurationOperation"), exports);
__exportStar(require("./ServerWide/Operations/SetDatabaseDynamicDistributionOperation"), exports);
__exportStar(require("./ServerWide/Operations/UpdateUnusedDatabasesOperation"), exports);
__exportStar(require("./ServerWide/Operations"), exports);
__exportStar(require("./ServerWide/Operations/DeleteDatabasesOperation"), exports);
__exportStar(require("./ServerWide/Operations/GetDatabaseNamesOperation"), exports);
__exportStar(require("./ServerWide/Operations/GetServerWideOperationStateOperation"), exports);
__exportStar(require("./ServerWide/Operations/ServerWideOperationCompletionAwaiter"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/CertificateMetadata"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/EditClientCertificateOperation"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/ReplaceClusterCertificateOperation"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/GetCertificateMetadataOperation"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/GetCertificatesMetadataOperation"), exports);
__exportStar(require("./ServerWide/Operations/Configuration/GetServerWideClientConfigurationOperation"), exports);
__exportStar(require("./ServerWide/Operations/Configuration/PutServerWideClientConfigurationOperation"), exports);
__exportStar(require("./ServerWide/Operations/Logs/GetLogsConfigurationResult"), exports);
__exportStar(require("./ServerWide/Operations/Logs/GetLogsConfigurationOperation"), exports);
__exportStar(require("./ServerWide/Operations/Logs/LogMode"), exports);
__exportStar(require("./ServerWide/Operations/Logs/SetLogsConfigurationOperation"), exports);
__exportStar(require("./ServerWide/Operations/Configuration/DeleteServerWideBackupConfigurationOperation"), exports);
__exportStar(require("./ServerWide/Operations/Configuration/GetServerWideClientConfigurationOperation"), exports);
__exportStar(require("./ServerWide/Operations/Configuration/GetServerWideBackupConfigurationsOperation"), exports);
__exportStar(require("./ServerWide/Operations/Configuration/PutServerWideBackupConfigurationOperation"), exports);
__exportStar(require("./ServerWide/Operations/Configuration/ServerWideBackupConfiguration"), exports);
__exportStar(require("./ServerWide/Operations/Configuration/DatabaseSettings"), exports);
__exportStar(require("./ServerWide/Operations/Configuration/GetDatabaseSettingsOperation"), exports);
__exportStar(require("./ServerWide/Operations/Configuration/PutDatabaseSettingsOperation"), exports);
var GetDatabaseTopologyCommand_1 = require("./ServerWide/Commands/GetDatabaseTopologyCommand");
Object.defineProperty(exports, "GetDatabaseTopologyCommand", { enumerable: true, get: function () { return GetDatabaseTopologyCommand_1.GetDatabaseTopologyCommand; } });
var GetClusterTopologyCommand_1 = require("./ServerWide/Commands/GetClusterTopologyCommand");
Object.defineProperty(exports, "GetClusterTopologyCommand", { enumerable: true, get: function () { return GetClusterTopologyCommand_1.GetClusterTopologyCommand; } });
var GetTcpInfoCommand_1 = require("./ServerWide/Commands/GetTcpInfoCommand");
Object.defineProperty(exports, "GetTcpInfoCommand", { enumerable: true, get: function () { return GetTcpInfoCommand_1.GetTcpInfoCommand; } });
var AddClusterNodeCommand_1 = require("./ServerWide/Commands/Cluster/AddClusterNodeCommand");
Object.defineProperty(exports, "AddClusterNodeCommand", { enumerable: true, get: function () { return AddClusterNodeCommand_1.AddClusterNodeCommand; } });
var CreateDatabaseOperation_1 = require("./ServerWide/Operations/CreateDatabaseOperation");
Object.defineProperty(exports, "CreateDatabaseOperation", { enumerable: true, get: function () { return CreateDatabaseOperation_1.CreateDatabaseOperation; } });
__exportStar(require("./ServerWide/Operations/ModifyConflictSolverOperation"), exports);
__exportStar(require("./Documents/Operations/Etl/ConnectionString"), exports);
var BulkInsertOperation_1 = require("./Documents/BulkInsertOperation");
Object.defineProperty(exports, "BulkInsertOperation", { enumerable: true, get: function () { return BulkInsertOperation_1.BulkInsertOperation; } });
__exportStar(require("./Documents/Operations/Backups/BackupConfiguration"), exports);
__exportStar(require("./Documents/Operations/Backups/BackupTaskType"), exports);
var DatabaseHealthCheckOperation_1 = require("./Documents/Operations/DatabaseHealthCheckOperation");
Object.defineProperty(exports, "DatabaseHealthCheckOperation", { enumerable: true, get: function () { return DatabaseHealthCheckOperation_1.DatabaseHealthCheckOperation; } });
var GetDetailedCollectionStatisticsOperation_1 = require("./Documents/Operations/GetDetailedCollectionStatisticsOperation");
Object.defineProperty(exports, "GetDetailedCollectionStatisticsOperation", { enumerable: true, get: function () { return GetDetailedCollectionStatisticsOperation_1.GetDetailedCollectionStatisticsOperation; } });
__exportStar(require("./Documents/Operations/OperationAbstractions"), exports);
var CompactDatabaseOperation_1 = require("./Documents/Operations/CompactDatabaseOperation");
Object.defineProperty(exports, "CompactDatabaseOperation", { enumerable: true, get: function () { return CompactDatabaseOperation_1.CompactDatabaseOperation; } });
var PutConnectionStringOperation_1 = require("./Documents/Operations/ConnectionStrings/PutConnectionStringOperation");
Object.defineProperty(exports, "PutConnectionStringOperation", { enumerable: true, get: function () { return PutConnectionStringOperation_1.PutConnectionStringOperation; } });
var PatchOperation_1 = require("./Documents/Operations/PatchOperation");
Object.defineProperty(exports, "PatchOperation", { enumerable: true, get: function () { return PatchOperation_1.PatchOperation; } });
var DeleteSorterOperation_1 = require("./Documents/Operations/Sorters/DeleteSorterOperation");
Object.defineProperty(exports, "DeleteSorterOperation", { enumerable: true, get: function () { return DeleteSorterOperation_1.DeleteSorterOperation; } });
var PutSortersOperation_1 = require("./Documents/Operations/Sorters/PutSortersOperation");
Object.defineProperty(exports, "PutSortersOperation", { enumerable: true, get: function () { return PutSortersOperation_1.PutSortersOperation; } });
var PatchByQueryOperation_1 = require("./Documents/Operations/PatchByQueryOperation");
Object.defineProperty(exports, "PatchByQueryOperation", { enumerable: true, get: function () { return PatchByQueryOperation_1.PatchByQueryOperation; } });
var PutCompareExchangeValueOperation_1 = require("./Documents/Operations/CompareExchange/PutCompareExchangeValueOperation");
Object.defineProperty(exports, "PutCompareExchangeValueOperation", { enumerable: true, get: function () { return PutCompareExchangeValueOperation_1.PutCompareExchangeValueOperation; } });
var GetCompareExchangeValueOperation_1 = require("./Documents/Operations/CompareExchange/GetCompareExchangeValueOperation");
Object.defineProperty(exports, "GetCompareExchangeValueOperation", { enumerable: true, get: function () { return GetCompareExchangeValueOperation_1.GetCompareExchangeValueOperation; } });
var CompareExchangeResult_1 = require("./Documents/Operations/CompareExchange/CompareExchangeResult");
Object.defineProperty(exports, "CompareExchangeResult", { enumerable: true, get: function () { return CompareExchangeResult_1.CompareExchangeResult; } });
var CompareExchangeValue_1 = require("./Documents/Operations/CompareExchange/CompareExchangeValue");
Object.defineProperty(exports, "CompareExchangeValue", { enumerable: true, get: function () { return CompareExchangeValue_1.CompareExchangeValue; } });
var CompareExchangeValueResultParser_1 = require("./Documents/Operations/CompareExchange/CompareExchangeValueResultParser");
Object.defineProperty(exports, "CompareExchangeValueResultParser", { enumerable: true, get: function () { return CompareExchangeValueResultParser_1.CompareExchangeValueResultParser; } });
var GetCompareExchangeValuesOperation_1 = require("./Documents/Operations/CompareExchange/GetCompareExchangeValuesOperation");
Object.defineProperty(exports, "GetCompareExchangeValuesOperation", { enumerable: true, get: function () { return GetCompareExchangeValuesOperation_1.GetCompareExchangeValuesOperation; } });
var DeleteCompareExchangeValueOperation_1 = require("./Documents/Operations/CompareExchange/DeleteCompareExchangeValueOperation");
Object.defineProperty(exports, "DeleteCompareExchangeValueOperation", { enumerable: true, get: function () { return DeleteCompareExchangeValueOperation_1.DeleteCompareExchangeValueOperation; } });
var CompareExchangeSessionValue_1 = require("./Documents/Operations/CompareExchange/CompareExchangeSessionValue");
Object.defineProperty(exports, "CompareExchangeSessionValue", { enumerable: true, get: function () { return CompareExchangeSessionValue_1.CompareExchangeSessionValue; } });
var CompareExchangeValueJsonConverter_1 = require("./Documents/Operations/CompareExchange/CompareExchangeValueJsonConverter");
Object.defineProperty(exports, "CompareExchangeValueJsonConverter", { enumerable: true, get: function () { return CompareExchangeValueJsonConverter_1.CompareExchangeValueJsonConverter; } });
var DeleteByQueryOperation_1 = require("./Documents/Operations/DeleteByQueryOperation");
Object.defineProperty(exports, "DeleteByQueryOperation", { enumerable: true, get: function () { return DeleteByQueryOperation_1.DeleteByQueryOperation; } });
var GetCollectionStatisticsOperation_1 = require("./Documents/Operations/GetCollectionStatisticsOperation");
Object.defineProperty(exports, "GetCollectionStatisticsOperation", { enumerable: true, get: function () { return GetCollectionStatisticsOperation_1.GetCollectionStatisticsOperation; } });
var GetNextOperationIdCommand_1 = require("./Documents/Commands/GetNextOperationIdCommand");
Object.defineProperty(exports, "GetNextOperationIdCommand", { enumerable: true, get: function () { return GetNextOperationIdCommand_1.GetNextOperationIdCommand; } });
var KillOperationCommand_1 = require("./Documents/Commands/KillOperationCommand");
Object.defineProperty(exports, "KillOperationCommand", { enumerable: true, get: function () { return KillOperationCommand_1.KillOperationCommand; } });
var DeleteDocumentCommand_1 = require("./Documents/Commands/DeleteDocumentCommand");
Object.defineProperty(exports, "DeleteDocumentCommand", { enumerable: true, get: function () { return DeleteDocumentCommand_1.DeleteDocumentCommand; } });
var NextIdentityForCommand_1 = require("./Documents/Commands/NextIdentityForCommand");
Object.defineProperty(exports, "NextIdentityForCommand", { enumerable: true, get: function () { return NextIdentityForCommand_1.NextIdentityForCommand; } });
var SeedIdentityForCommand_1 = require("./Documents/Commands/SeedIdentityForCommand");
Object.defineProperty(exports, "SeedIdentityForCommand", { enumerable: true, get: function () { return SeedIdentityForCommand_1.SeedIdentityForCommand; } });
var ExplainQueryCommand_1 = require("./Documents/Commands/ExplainQueryCommand");
Object.defineProperty(exports, "ExplainQueryCommand", { enumerable: true, get: function () { return ExplainQueryCommand_1.ExplainQueryCommand; } });
var GetIdentitiesOperation_1 = require("./Documents/Operations/Identities/GetIdentitiesOperation");
Object.defineProperty(exports, "GetIdentitiesOperation", { enumerable: true, get: function () { return GetIdentitiesOperation_1.GetIdentitiesOperation; } });
var GetStatisticsOperation_1 = require("./Documents/Operations/GetStatisticsOperation");
Object.defineProperty(exports, "GetStatisticsOperation", { enumerable: true, get: function () { return GetStatisticsOperation_1.GetStatisticsOperation; } });
Object.defineProperty(exports, "GetStatisticsCommand", { enumerable: true, get: function () { return GetStatisticsOperation_1.GetStatisticsCommand; } });
var GetOperationStateOperation_1 = require("./Documents/Operations/GetOperationStateOperation");
Object.defineProperty(exports, "GetOperationStateOperation", { enumerable: true, get: function () { return GetOperationStateOperation_1.GetOperationStateOperation; } });
var MaintenanceOperationExecutor_1 = require("./Documents/Operations/MaintenanceOperationExecutor");
Object.defineProperty(exports, "MaintenanceOperationExecutor", { enumerable: true, get: function () { return MaintenanceOperationExecutor_1.MaintenanceOperationExecutor; } });
var OperationCompletionAwaiter_1 = require("./Documents/Operations/OperationCompletionAwaiter");
Object.defineProperty(exports, "OperationCompletionAwaiter", { enumerable: true, get: function () { return OperationCompletionAwaiter_1.OperationCompletionAwaiter; } });
var GetClientConfigurationOperation_1 = require("./Documents/Operations/Configuration/GetClientConfigurationOperation");
Object.defineProperty(exports, "GetClientConfigurationOperation", { enumerable: true, get: function () { return GetClientConfigurationOperation_1.GetClientConfigurationOperation; } });
var PutClientConfigurationOperation_1 = require("./Documents/Operations/Configuration/PutClientConfigurationOperation");
Object.defineProperty(exports, "PutClientConfigurationOperation", { enumerable: true, get: function () { return PutClientConfigurationOperation_1.PutClientConfigurationOperation; } });
var PutDocumentCommand_1 = require("./Documents/Commands/PutDocumentCommand");
Object.defineProperty(exports, "PutDocumentCommand", { enumerable: true, get: function () { return PutDocumentCommand_1.PutDocumentCommand; } });
var GetIndexNamesOperation_1 = require("./Documents/Operations/Indexes/GetIndexNamesOperation");
Object.defineProperty(exports, "GetIndexNamesOperation", { enumerable: true, get: function () { return GetIndexNamesOperation_1.GetIndexNamesOperation; } });
var DeleteIndexErrorsOperation_1 = require("./Documents/Operations/Indexes/DeleteIndexErrorsOperation");
Object.defineProperty(exports, "DeleteIndexErrorsOperation", { enumerable: true, get: function () { return DeleteIndexErrorsOperation_1.DeleteIndexErrorsOperation; } });
var DisableIndexOperation_1 = require("./Documents/Operations/Indexes/DisableIndexOperation");
Object.defineProperty(exports, "DisableIndexOperation", { enumerable: true, get: function () { return DisableIndexOperation_1.DisableIndexOperation; } });
var EnableIndexOperation_1 = require("./Documents/Operations/Indexes/EnableIndexOperation");
Object.defineProperty(exports, "EnableIndexOperation", { enumerable: true, get: function () { return EnableIndexOperation_1.EnableIndexOperation; } });
var GetIndexingStatusOperation_1 = require("./Documents/Operations/Indexes/GetIndexingStatusOperation");
Object.defineProperty(exports, "GetIndexingStatusOperation", { enumerable: true, get: function () { return GetIndexingStatusOperation_1.GetIndexingStatusOperation; } });
var GetIndexesStatisticsOperation_1 = require("./Documents/Operations/Indexes/GetIndexesStatisticsOperation");
Object.defineProperty(exports, "GetIndexesStatisticsOperation", { enumerable: true, get: function () { return GetIndexesStatisticsOperation_1.GetIndexesStatisticsOperation; } });
var GetIndexStatisticsOperation_1 = require("./Documents/Operations/Indexes/GetIndexStatisticsOperation");
Object.defineProperty(exports, "GetIndexStatisticsOperation", { enumerable: true, get: function () { return GetIndexStatisticsOperation_1.GetIndexStatisticsOperation; } });
var GetIndexesOperation_1 = require("./Documents/Operations/Indexes/GetIndexesOperation");
Object.defineProperty(exports, "GetIndexesOperation", { enumerable: true, get: function () { return GetIndexesOperation_1.GetIndexesOperation; } });
var GetTermsOperation_1 = require("./Documents/Operations/Indexes/GetTermsOperation");
Object.defineProperty(exports, "GetTermsOperation", { enumerable: true, get: function () { return GetTermsOperation_1.GetTermsOperation; } });
var IndexHasChangedOperation_1 = require("./Documents/Operations/Indexes/IndexHasChangedOperation");
Object.defineProperty(exports, "IndexHasChangedOperation", { enumerable: true, get: function () { return IndexHasChangedOperation_1.IndexHasChangedOperation; } });
var PutIndexesOperation_1 = require("./Documents/Operations/Indexes/PutIndexesOperation");
Object.defineProperty(exports, "PutIndexesOperation", { enumerable: true, get: function () { return PutIndexesOperation_1.PutIndexesOperation; } });
var StopIndexingOperation_1 = require("./Documents/Operations/Indexes/StopIndexingOperation");
Object.defineProperty(exports, "StopIndexingOperation", { enumerable: true, get: function () { return StopIndexingOperation_1.StopIndexingOperation; } });
var StartIndexingOperation_1 = require("./Documents/Operations/Indexes/StartIndexingOperation");
Object.defineProperty(exports, "StartIndexingOperation", { enumerable: true, get: function () { return StartIndexingOperation_1.StartIndexingOperation; } });
var StopIndexOperation_1 = require("./Documents/Operations/Indexes/StopIndexOperation");
Object.defineProperty(exports, "StopIndexOperation", { enumerable: true, get: function () { return StopIndexOperation_1.StopIndexOperation; } });
var StartIndexOperation_1 = require("./Documents/Operations/Indexes/StartIndexOperation");
Object.defineProperty(exports, "StartIndexOperation", { enumerable: true, get: function () { return StartIndexOperation_1.StartIndexOperation; } });
var ResetIndexOperation_1 = require("./Documents/Operations/Indexes/ResetIndexOperation");
Object.defineProperty(exports, "ResetIndexOperation", { enumerable: true, get: function () { return ResetIndexOperation_1.ResetIndexOperation; } });
var DeleteIndexOperation_1 = require("./Documents/Operations/Indexes/DeleteIndexOperation");
Object.defineProperty(exports, "DeleteIndexOperation", { enumerable: true, get: function () { return DeleteIndexOperation_1.DeleteIndexOperation; } });
var NextIdentityForOperation_1 = require("./Documents/Operations/Identities/NextIdentityForOperation");
Object.defineProperty(exports, "NextIdentityForOperation", { enumerable: true, get: function () { return NextIdentityForOperation_1.NextIdentityForOperation; } });
var SeedIdentityForOperation_1 = require("./Documents/Operations/Identities/SeedIdentityForOperation");
Object.defineProperty(exports, "SeedIdentityForOperation", { enumerable: true, get: function () { return SeedIdentityForOperation_1.SeedIdentityForOperation; } });
var UpdateExternalReplicationOperation_1 = require("./Documents/Operations/Replication/UpdateExternalReplicationOperation");
Object.defineProperty(exports, "UpdateExternalReplicationOperation", { enumerable: true, get: function () { return UpdateExternalReplicationOperation_1.UpdateExternalReplicationOperation; } });
var PutPullReplicationAsHubOperation_1 = require("./Documents/Operations/Replication/PutPullReplicationAsHubOperation");
Object.defineProperty(exports, "PutPullReplicationAsHubOperation", { enumerable: true, get: function () { return PutPullReplicationAsHubOperation_1.PutPullReplicationAsHubOperation; } });
__exportStar(require("./Documents/Operations/Replication/DetailedReplicationHubAccess"), exports);
__exportStar(require("./Documents/Operations/Replication/GetReplicationHubAccessOperation"), exports);
__exportStar(require("./Documents/Operations/Replication/IExternalReplication"), exports);
__exportStar(require("./Documents/Operations/Replication/PreventDeletionsMode"), exports);
__exportStar(require("./Documents/Operations/Replication/PullReplicationMode"), exports);
__exportStar(require("./Documents/Operations/Replication/RegisterReplicationHubAccessOperation"), exports);
__exportStar(require("./Documents/Operations/Replication/ReplicationHubAccess"), exports);
__exportStar(require("./Documents/Operations/Replication/ReplicationHubAccessResult"), exports);
__exportStar(require("./Documents/Operations/Replication/ReplicationHubAccessResponse"), exports);
__exportStar(require("./Documents/Operations/Replication/UnregisterReplicationHubAccessOperation"), exports);
var UpdatePullReplicationAsSinkOperation_1 = require("./Documents/Operations/Replication/UpdatePullReplicationAsSinkOperation");
Object.defineProperty(exports, "UpdatePullReplicationAsSinkOperation", { enumerable: true, get: function () { return UpdatePullReplicationAsSinkOperation_1.UpdatePullReplicationAsSinkOperation; } });
var GetConflictsCommand_1 = require("./Documents/Commands/GetConflictsCommand");
Object.defineProperty(exports, "GetConflictsCommand", { enumerable: true, get: function () { return GetConflictsCommand_1.GetConflictsCommand; } });
var SetIndexesLockOperation_1 = require("./Documents/Operations/Indexes/SetIndexesLockOperation");
Object.defineProperty(exports, "SetIndexesLockOperation", { enumerable: true, get: function () { return SetIndexesLockOperation_1.SetIndexesLockOperation; } });
var SetIndexesPriorityOperation_1 = require("./Documents/Operations/Indexes/SetIndexesPriorityOperation");
Object.defineProperty(exports, "SetIndexesPriorityOperation", { enumerable: true, get: function () { return SetIndexesPriorityOperation_1.SetIndexesPriorityOperation; } });
__exportStar(require("./Documents/Operations/PatchRequest"), exports);
__exportStar(require("./Documents/Operations/GetDetailedStatisticsOperation"), exports);
__exportStar(require("./Documents/Commands/Batches/BatchOptions"), exports);
__exportStar(require("./Documents/Commands/Batches/DeleteAttachmentCommandData"), exports);
__exportStar(require("./Documents/Commands/Batches/PatchCommandData"), exports);
__exportStar(require("./Documents/Commands/Batches/PutAttachmentCommandData"), exports);
__exportStar(require("./Documents/Commands/Batches/PutAttachmentCommandHelper"), exports);
__exportStar(require("./Documents/Commands/CommandData"), exports);
__exportStar(require("./ServerWide/Operations/GetDatabaseRecordOperation"), exports);
__exportStar(require("./Documents/SetupDocumentBase"), exports);
__exportStar(require("./Documents/Commands/StreamResultResponse"), exports);
__exportStar(require("./Documents/Commands/StreamResult"), exports);
__exportStar(require("./Documents/Session/Operations/BatchOperation"), exports);
__exportStar(require("./Documents/Session/Operations/GetRevisionOperation"), exports);
__exportStar(require("./Documents/Session/Operations/GetRevisionsCountOperation"), exports);
__exportStar(require("./Documents/Lazy"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/IEagerSessionOperations"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/ILazyOperation"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/ILazySessionOperations"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/LazyAggregationQueryOperation"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/LazyLoadOperation"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/LazyQueryOperation"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/LazySessionOperations"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/LazyStartsWithOperation"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/LazySuggestionQueryOperation"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/LazyClusterTransactionOperations"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/LazyGetCompareExchangeValueOperation"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/LazyGetCompareExchangeValuesOperation"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/LazyConditionalLoadOperation"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/LazyRevisionOperation"), exports);
__exportStar(require("./Documents/Session/Operations/Lazy/LazyRevisionOperations"), exports);
__exportStar(require("./Documents/Session/Operations/LoadOperation"), exports);
__exportStar(require("./Documents/Session/Operations/LoadStartingWithOperation"), exports);
__exportStar(require("./Documents/Session/Operations/MultiGetOperation"), exports);
__exportStar(require("./Documents/Session/Operations/QueryOperation"), exports);
__exportStar(require("./Documents/Session/Operations/StreamOperation"), exports);
__exportStar(require("./Documents/Operations/Attachments/DeleteAttachmentOperation"), exports);
__exportStar(require("./Documents/Operations/Attachments/PutAttachmentOperation"), exports);
__exportStar(require("./Documents/Operations/PatchResult"), exports);
__exportStar(require("./Documents/Operations/PatchStatus"), exports);
__exportStar(require("./Documents/Operations/Revisions/ConfigureRevisionsOperation"), exports);
__exportStar(require("./Documents/Operations/Revisions/GetRevisionsOperation"), exports);
__exportStar(require("./Documents/Operations/Revisions/RevisionsResult"), exports);
__exportStar(require("./Documents/Operations/RevisionsCollectionConfiguration"), exports);
__exportStar(require("./Documents/Operations/RevisionsConfiguration"), exports);
__exportStar(require("./Documents/Operations/DetailedDatabaseStatistics"), exports);
__exportStar(require("./Documents/Operations/SessionOperationExecutor"), exports);
__exportStar(require("./Documents/Operations/Configuration/StudioConfiguration"), exports);
__exportStar(require("./Documents/Operations/Configuration/StudioEnvironment"), exports);
__exportStar(require("./Documents/Operations/ConnectionStrings/GetConnectionStringsOperation"), exports);
__exportStar(require("./Documents/Operations/ConnectionStrings/RemoveConnectionStringOperation"), exports);
__exportStar(require("./Documents/Operations/Etl/EtlConfiguration"), exports);
__exportStar(require("./Documents/Operations/Etl/RavenEtlConfiguration"), exports);
__exportStar(require("./Documents/Operations/Etl/Sql/SqlEtlConfiguration"), exports);
__exportStar(require("./Documents/Operations/Etl/Sql/SqlEtlTable"), exports);
__exportStar(require("./Documents/Operations/Etl/Olap/OlapEtlConfiguration"), exports);
__exportStar(require("./Documents/Operations/Etl/Olap/OlapEtlFileFormat"), exports);
__exportStar(require("./Documents/Operations/Etl/Olap/OlapEtlTable"), exports);
__exportStar(require("./Documents/Operations/Etl/Transformation"), exports);
__exportStar(require("./Documents/Operations/Expiration/ExpirationConfiguration"), exports);
__exportStar(require("./Documents/Operations/Replication/PullReplicationAsSink"), exports);
__exportStar(require("./Documents/Operations/Replication/PullReplicationDefinition"), exports);
__exportStar(require("./Documents/Operations/Etl/AddEtlOperation"), exports);
__exportStar(require("./Documents/Operations/Etl/UpdateEtlOperation"), exports);
__exportStar(require("./Documents/Operations/Etl/ResetEtlOperation"), exports);
__exportStar(require("./Documents/Operations/DisableDatabaseToggleResult"), exports);
__exportStar(require("./Documents/Operations/Expiration/ConfigureExpirationOperation"), exports);
__exportStar(require("./Documents/Operations/OngoingTasks/DeleteOngoingTaskOperation"), exports);
__exportStar(require("./Documents/Operations/OngoingTasks/GetPullReplicationHubTasksInfoOperation"), exports);
__exportStar(require("./Documents/Operations/OngoingTasks/OngoingTaskPullReplicationAsSink"), exports);
__exportStar(require("./Documents/Operations/OngoingTasks/OngoingTaskPullReplicationAsHub"), exports);
__exportStar(require("./Documents/Operations/OngoingTasks/OngoingTaskType"), exports);
__exportStar(require("./Documents/Operations/OngoingTasks/RunningBackup"), exports);
__exportStar(require("./Documents/Operations/OngoingTasks/NextBackup"), exports);
__exportStar(require("./Documents/Operations/GetOngoingTaskInfoOperation"), exports);
__exportStar(require("./Documents/Operations/OngoingTasks/ToggleOngoingTaskStateOperation"), exports);
__exportStar(require("./Documents/Operations/Refresh/ConfigureRefreshOperation"), exports);
__exportStar(require("./Documents/Operations/Refresh/RefreshConfiguration"), exports);
__exportStar(require("./Documents/Operations/Refresh/ConfigureRefreshOperationResult"), exports);
__exportStar(require("./Documents/Operations/ToggleDatabasesStateOperation"), exports);
__exportStar(require("./Documents/Operations/TransactionsRecording/StartTransactionsRecordingOperation"), exports);
__exportStar(require("./Documents/Operations/TransactionsRecording/StopTransactionsRecordingOperation"), exports);
__exportStar(require("./Documents/Operations/Backups/AmazonSettings"), exports);
__exportStar(require("./Documents/Operations/Backups/AzureSettings"), exports);
__exportStar(require("./Documents/Operations/Backups/BackupEncryptionSettings"), exports);
__exportStar(require("./Documents/Operations/Backups/BackupEncryptionSettings"), exports);
__exportStar(require("./Documents/Operations/Backups/Enums"), exports);
__exportStar(require("./Documents/Operations/Backups/FtpSettings"), exports);
__exportStar(require("./Documents/Operations/Backups/GlacierSettings"), exports);
__exportStar(require("./Documents/Operations/Backups/LocalSettings"), exports);
__exportStar(require("./Documents/Operations/Backups/PeriodicBackupConfiguration"), exports);
__exportStar(require("./Documents/Operations/Backups/S3Settings"), exports);
__exportStar(require("./Documents/Operations/Backups/BackupSettings"), exports);
__exportStar(require("./Documents/Operations/Backups/BackupStatus"), exports);
__exportStar(require("./Documents/Operations/Backups/GetPeriodicBackupStatusOperation"), exports);
__exportStar(require("./Documents/Operations/Backups/GetPeriodicBackupStatusOperationResult"), exports);
__exportStar(require("./Documents/Operations/Backups/LastRaftIndex"), exports);
__exportStar(require("./Documents/Operations/Backups/PeriodicBackupStatus"), exports);
__exportStar(require("./Documents/Operations/Backups/RestoreBackupConfiguration"), exports);
__exportStar(require("./Documents/Operations/Backups/RestoreBackupOperation"), exports);
__exportStar(require("./Documents/Operations/Backups/StartBackupOperation"), exports);
__exportStar(require("./Documents/Operations/Backups/StartBackupOperationResult"), exports);
__exportStar(require("./Documents/Operations/Backups/UpdatePeriodicBackupOperation"), exports);
__exportStar(require("./Documents/Operations/Backups/UpdatePeriodicBackupOperationResult"), exports);
__exportStar(require("./Documents/Operations/Backups/UploadProgress"), exports);
__exportStar(require("./Documents/Operations/Backups/UploadState"), exports);
__exportStar(require("./Documents/Operations/Backups/CompressionLevel"), exports);
__exportStar(require("./Documents/Operations/Backups/GetBackupConfigurationScript"), exports);
__exportStar(require("./Documents/Operations/Backups/GoogleCloudSettings"), exports);
__exportStar(require("./Documents/Operations/Backups/RestoreBackupConfigurationBase"), exports);
__exportStar(require("./Documents/Operations/Backups/RestoreFromAzureConfiguration"), exports);
__exportStar(require("./Documents/Operations/Backups/RestoreFromGoogleCloudConfiguration"), exports);
__exportStar(require("./Documents/Operations/Backups/RestoreFromS3Configuration"), exports);
__exportStar(require("./Documents/Operations/Backups/RestoreType"), exports);
__exportStar(require("./Documents/Operations/Backups/RetentionPolicy"), exports);
var GetIndexOperation_1 = require("./Documents/Operations/Indexes/GetIndexOperation");
Object.defineProperty(exports, "GetIndexOperation", { enumerable: true, get: function () { return GetIndexOperation_1.GetIndexOperation; } });
var GetIndexErrorsOperation_1 = require("./Documents/Operations/Indexes/GetIndexErrorsOperation");
Object.defineProperty(exports, "GetIndexErrorsOperation", { enumerable: true, get: function () { return GetIndexErrorsOperation_1.GetIndexErrorsOperation; } });
__exportStar(require("./Documents/Indexes/Enums"), exports);
__exportStar(require("./Documents/Indexes/IndexDeploymentMode"), exports);
__exportStar(require("./Documents/Indexes/IndexDefinition"), exports);
__exportStar(require("./Documents/Indexes/AbstractCommonApiForIndexes"), exports);
__exportStar(require("./Documents/Indexes/AbstractIndexDefinitionBuilder"), exports);
__exportStar(require("./Documents/Indexes/IAbstractIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/Errors"), exports);
__exportStar(require("./Documents/Indexes/AdditionalAssembly"), exports);
__exportStar(require("./Documents/Indexes/IndexDefinitionHelper"), exports);
__exportStar(require("./Documents/Indexes/IndexFieldOptions"), exports);
__exportStar(require("./Documents/Indexes/Spatial"), exports);
__exportStar(require("./Documents/Indexes/IndexingStatus"), exports);
__exportStar(require("./Documents/Indexes/RollingIndex"), exports);
__exportStar(require("./Documents/Indexes/RollingIndexDeployment"), exports);
__exportStar(require("./Documents/Indexes/RollingIndexState"), exports);
__exportStar(require("./Documents/Indexes/IndexStats"), exports);
__exportStar(require("./Documents/Indexes/IndexSourceType"), exports);
__exportStar(require("./Documents/Indexes"), exports);
__exportStar(require("./Documents/Indexes/StronglyTyped"), exports);
__exportStar(require("./Documents/Indexes/IndexDefinitionBase"), exports);
__exportStar(require("./Documents/Indexes/Analysis/AnalyzerDefinition"), exports);
__exportStar(require("./Documents/Indexes/AbstractCsharpIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/AbstractCsharpMultiMapIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/AbstractJavaScriptIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/BaseJavaScriptIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/AbstractJavaScriptMultiMapIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/AbstractRawJavaScriptIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/AutoIndexDefinition"), exports);
__exportStar(require("./Documents/Indexes/AutoIndexFieldOptions"), exports);
__exportStar(require("./Documents/Indexes/Spatial/AutoSpatialOptions"), exports);
__exportStar(require("./Documents/Indexes/Counters/AbstractCountersIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/Counters/AbstractGenericCountersIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/Counters/AbstractCsharpCountersIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/Counters/AbstractMultiMapCountersIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/Counters/AbstractRawJavaScriptCountersIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/Counters/CountersIndexDefinition"), exports);
__exportStar(require("./Documents/Indexes/Counters/CountersIndexDefinitionBuilder"), exports);
__exportStar(require("./Documents/Indexes/TimeSeries/AbstractGenericTimeSeriesIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/TimeSeries/AbstractMultiMapTimeSeriesIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/TimeSeries/AbstractCsharpTimeSeriesIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/TimeSeries/AbstractRawJavaScriptTimeSeriesIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/TimeSeries/AbstractTimeSeriesIndexCreationTask"), exports);
__exportStar(require("./Documents/Indexes/TimeSeries/TimeSeriesIndexDefinition"), exports);
__exportStar(require("./Documents/Indexes/TimeSeries/TimeSeriesIndexDefinitionBuilder"), exports);
__exportStar(require("./Documents/Replication/ExternalReplication"), exports);
__exportStar(require("./Documents/Replication/ReplicationNode"), exports);
__exportStar(require("./Documents/Replication/ExternalReplicationBase"), exports);
__exportStar(require("./Documents/DocumentAbstractions"), exports);
__exportStar(require("./Documents/DocumentStore"), exports);
__exportStar(require("./Documents/DocumentStoreBase"), exports);
__exportStar(require("./Documents/IDocumentStore"), exports);
__exportStar(require("./Documents/IdTypeAndName"), exports);
__exportStar(require("./Documents/Subscriptions/SubscriptionBatch"), exports);
__exportStar(require("./Documents/Subscriptions/DocumentSubscriptions"), exports);
__exportStar(require("./Documents/Subscriptions/SubscriptionWorker"), exports);
__exportStar(require("./Documents/Subscriptions/SubscriptionWorkerOptions"), exports);
__exportStar(require("./Documents/Subscriptions/SubscriptionCreationOptions"), exports);
__exportStar(require("./Documents/Subscriptions/Revision"), exports);
__exportStar(require("./Documents/Subscriptions/SubscriptionState"), exports);
__exportStar(require("./Documents/Subscriptions/SubscriptionCreationOptions"), exports);
__exportStar(require("./Documents/Subscriptions/UpdateSubscriptionResult"), exports);
__exportStar(require("./Documents/Subscriptions/SubscriptionOpeningStrategy"), exports);
__exportStar(require("./Documents/Subscriptions/SubscriptionUpdateOptions"), exports);
__exportStar(require("./Documents/Session/AbstractDocumentQuery"), exports);
__exportStar(require("./Documents/Session/CmpXchg"), exports);
__exportStar(require("./Documents/Session/DocumentInfo"), exports);
__exportStar(require("./Documents/Session/DocumentQuery"), exports);
__exportStar(require("./Documents/Session/DocumentQueryHelper"), exports);
__exportStar(require("./Documents/Session/DocumentsById"), exports);
__exportStar(require("./Documents/Session/DocumentsChanges"), exports);
__exportStar(require("./Documents/Session/DocumentSession"), exports);
__exportStar(require("./Documents/Session/EntityToJson"), exports);
__exportStar(require("./Documents/Session/ForceRevisionStrategy"), exports);
__exportStar(require("./Documents/Session/GraphDocumentQuery"), exports);
__exportStar(require("./Documents/Session/GroupByDocumentQuery"), exports);
__exportStar(require("./Documents/Session/GroupByField"), exports);
__exportStar(require("./Documents/Session/IAbstractDocumentQuery"), exports);
__exportStar(require("./Documents/Session/IAbstractDocumentQueryImpl"), exports);
__exportStar(require("./Documents/Session/ILazyRevisionsOperations"), exports);
__exportStar(require("./Documents/Session/IAdvancedSessionOperations"), exports);
__exportStar(require("./Documents/Session/IDocumentQuery"), exports);
__exportStar(require("./Documents/Session/IDocumentQueryBase"), exports);
__exportStar(require("./Documents/Session/IDocumentQueryBuilder"), exports);
__exportStar(require("./Documents/Session/IDocumentQueryBaseSingle"), exports);
__exportStar(require("./Documents/Session/IDocumentSession"), exports);
__exportStar(require("./Documents/Session/IEnumerableQuery"), exports);
__exportStar(require("./Documents/Session/IFilterDocumentQueryBase"), exports);
__exportStar(require("./Documents/Session/IGraphDocumentQuery"), exports);
__exportStar(require("./Documents/Session/IGroupByDocumentQuery"), exports);
__exportStar(require("./Documents/Session/IncludesUtil"), exports);
__exportStar(require("./Documents/Session/InMemoryDocumentSessionOperations"), exports);
__exportStar(require("./Documents/Session/IQueryBase"), exports);
__exportStar(require("./Documents/Session/IRawDocumentQuery"), exports);
__exportStar(require("./Documents/Session/MethodCall"), exports);
__exportStar(require("./Documents/Session/OrderingType"), exports);
__exportStar(require("./Documents/Session/QueryEvents"), exports);
__exportStar(require("./Documents/Session/QueryOptions"), exports);
__exportStar(require("./Documents/Session/QueryStatistics"), exports);
__exportStar(require("./Documents/Session/StreamQueryStatistics"), exports);
__exportStar(require("./Documents/Session/RawDocumentQuery"), exports);
__exportStar(require("./Documents/Session/SessionEvents"), exports);
__exportStar(require("./Documents/Session/WhereParams"), exports);
__exportStar(require("./Documents/Session/ILazyClusterTransactionOperations"), exports);
__exportStar(require("./Documents/Session/ISessionDocumentAppendTimeSeriesBase"), exports);
__exportStar(require("./Documents/Session/ISessionDocumentDeleteTimeSeriesBase"), exports);
__exportStar(require("./Documents/Session/ISessionDocumentRollupTypedAppendTimeSeriesBase"), exports);
__exportStar(require("./Documents/Session/ISessionDocumentRollupTypedTimeSeries"), exports);
__exportStar(require("./Documents/Session/ISessionDocumentTimeSeries"), exports);
__exportStar(require("./Documents/Session/ISessionDocumentTypedAppendTimeSeriesBase"), exports);
__exportStar(require("./Documents/Session/ISessionDocumentTypedTimeSeries"), exports);
__exportStar(require("./Documents/Session/JavaScriptMap"), exports);
__exportStar(require("./Documents/Session/IMetadataDictionary"), exports);
__exportStar(require("./Documents/Session/DocumentResultStream"), exports);
__exportStar(require("./Documents/Session/SessionDocumentRollupTypedTimeSeries"), exports);
__exportStar(require("./Documents/Session/SessionDocumentTimeSeries"), exports);
__exportStar(require("./Documents/Session/SessionDocumentTypedTimeSeries"), exports);
__exportStar(require("./Documents/Session/SessionTimeSeriesBase"), exports);
__exportStar(require("./Documents/Session/Loaders/ICounterIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/IAbstractTimeSeriesIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/ICompareExchangeValueIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/IDocumentIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/IGenericIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/IGenericRevisionIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/IGenericTimeSeriesIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/ISubscriptionIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/ISubscriptionTimeSeriesIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/TimeSeriesIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/SubscriptionIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/ILazyLoaderWithInclude"), exports);
__exportStar(require("./Documents/Session/Loaders/ILoaderWithInclude"), exports);
__exportStar(require("./Documents/Session/Loaders/ITimeSeriesIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/LazyMultiLoaderWithInclude"), exports);
__exportStar(require("./Documents/Session/Loaders/MultiLoaderWithInclude"), exports);
__exportStar(require("./Documents/Session/DocumentQueryCustomization"), exports);
__exportStar(require("./Documents/Session/DocumentSessionAttachments"), exports);
__exportStar(require("./Documents/Session/DocumentSessionAttachmentsBase"), exports);
__exportStar(require("./Documents/Session/DocumentSessionRevisions"), exports);
__exportStar(require("./Documents/Session/DocumentSessionRevisionsBase"), exports);
__exportStar(require("./Documents/Session/IAttachmentsSessionOperations"), exports);
__exportStar(require("./Documents/Session/IDocumentQueryCustomization"), exports);
__exportStar(require("./Documents/Session/IRevisionsSessionOperations"), exports);
__exportStar(require("./Documents/Session/ResponseTimeInformation"), exports);
__exportStar(require("./Documents/Session/MetadataObject"), exports);
__exportStar(require("./Documents/Session/TransactionMode"), exports);
__exportStar(require("./Documents/Session/ConditionalLoadResult"), exports);
__exportStar(require("./Documents/Session/IClusterTransactionOperations"), exports);
__exportStar(require("./Documents/Session/ISessionDocumentCounters"), exports);
__exportStar(require("./Documents/Session/ClusterTransactionOperations"), exports);
__exportStar(require("./Documents/Session/CounterInternalTypes"), exports);
__exportStar(require("./Documents/Session/Loaders/IIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/IncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/IncludeBuilderBase"), exports);
__exportStar(require("./Documents/Session/Loaders/IQueryIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/QueryIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Loaders/QueryIncludeBuilder"), exports);
__exportStar(require("./Documents/Session/Operations/BatchCommandResult"), exports);
__exportStar(require("./Documents/Session/SessionDocumentCounters"), exports);
__exportStar(require("./Documents/Session/TimeSeries/TimeSeriesEntry"), exports);
__exportStar(require("./Documents/Session/TimeSeries/TimeSeriesValue"), exports);
__exportStar(require("./Documents/Session/TimeSeries/TimeSeriesValuesHelper"), exports);
__exportStar(require("./Documents/Session/TimeSeries/TypedTimeSeriesEntry"), exports);
__exportStar(require("./Documents/Session/TimeSeries/TypedTimeSeriesRollupEntry"), exports);
__exportStar(require("./Documents/TimeSeries/TimeSeriesOperations"), exports);
__exportStar(require("./Documents/Commands/StreamResult"), exports);
__exportStar(require("./Documents/Session/SessionOptions"), exports);
__exportStar(require("./Documents/Commands/CommandData"), exports);
__exportStar(require("./Documents/Commands/Batches/CopyAttachmentCommandData"), exports);
__exportStar(require("./Documents/Commands/Batches/DeleteAttachmentCommandData"), exports);
__exportStar(require("./Documents/Commands/Batches/MoveAttachmentCommandData"), exports);
__exportStar(require("./Documents/Commands/Batches/PutAttachmentCommandData"), exports);
__exportStar(require("./Documents/Commands/Batches/BatchPatchCommandData"), exports);
__exportStar(require("./Documents/Commands/Batches/CountersBatchCommandData"), exports);
__exportStar(require("./Documents/Commands/Batches/PatchCommandData"), exports);
__exportStar(require("./Documents/Commands/Batches/PutCompareExchangeCommandData"), exports);
__exportStar(require("./Documents/Commands/Batches/DeleteCompareExchangeCommandData"), exports);
__exportStar(require("./Documents/Lazy"), exports);
var CounterBatch_1 = require("./Documents/Operations/Counters/CounterBatch");
Object.defineProperty(exports, "CounterBatch", { enumerable: true, get: function () { return CounterBatch_1.CounterBatch; } });
var GetCountersOperation_1 = require("./Documents/Operations/Counters/GetCountersOperation");
Object.defineProperty(exports, "GetCountersOperation", { enumerable: true, get: function () { return GetCountersOperation_1.GetCountersOperation; } });
var CounterBatchOperation_1 = require("./Documents/Operations/Counters/CounterBatchOperation");
Object.defineProperty(exports, "CounterBatchOperation", { enumerable: true, get: function () { return CounterBatchOperation_1.CounterBatchOperation; } });
var CounterOperation_1 = require("./Documents/Operations/Counters/CounterOperation");
Object.defineProperty(exports, "CounterOperation", { enumerable: true, get: function () { return CounterOperation_1.CounterOperation; } });
var DocumentCountersOperation_1 = require("./Documents/Operations/Counters/DocumentCountersOperation");
Object.defineProperty(exports, "DocumentCountersOperation", { enumerable: true, get: function () { return DocumentCountersOperation_1.DocumentCountersOperation; } });
__exportStar(require("./Documents/Operations/Counters/CounterDetail"), exports);
__exportStar(require("./Documents/Operations/Counters/CountersDetail"), exports);
var RawTimeSeriesTypes_1 = require("./Documents/Operations/TimeSeries/RawTimeSeriesTypes");
Object.defineProperty(exports, "TIME_SERIES_ROLLUP_SEPARATOR", { enumerable: true, get: function () { return RawTimeSeriesTypes_1.TIME_SERIES_ROLLUP_SEPARATOR; } });
var ConfigureRawTimeSeriesPolicyOperation_1 = require("./Documents/Operations/TimeSeries/ConfigureRawTimeSeriesPolicyOperation");
Object.defineProperty(exports, "ConfigureRawTimeSeriesPolicyOperation", { enumerable: true, get: function () { return ConfigureRawTimeSeriesPolicyOperation_1.ConfigureRawTimeSeriesPolicyOperation; } });
var ConfigureTimeSeriesOperation_1 = require("./Documents/Operations/TimeSeries/ConfigureTimeSeriesOperation");
Object.defineProperty(exports, "ConfigureTimeSeriesOperation", { enumerable: true, get: function () { return ConfigureTimeSeriesOperation_1.ConfigureTimeSeriesOperation; } });
var ConfigureTimeSeriesPolicyOperation_1 = require("./Documents/Operations/TimeSeries/ConfigureTimeSeriesPolicyOperation");
Object.defineProperty(exports, "ConfigureTimeSeriesPolicyOperation", { enumerable: true, get: function () { return ConfigureTimeSeriesPolicyOperation_1.ConfigureTimeSeriesPolicyOperation; } });
var ConfigureTimeSeriesValueNamesOperation_1 = require("./Documents/Operations/TimeSeries/ConfigureTimeSeriesValueNamesOperation");
Object.defineProperty(exports, "ConfigureTimeSeriesValueNamesOperation", { enumerable: true, get: function () { return ConfigureTimeSeriesValueNamesOperation_1.ConfigureTimeSeriesValueNamesOperation; } });
var GetMultipleTimeSeriesOperation_1 = require("./Documents/Operations/TimeSeries/GetMultipleTimeSeriesOperation");
Object.defineProperty(exports, "GetMultipleTimeSeriesOperation", { enumerable: true, get: function () { return GetMultipleTimeSeriesOperation_1.GetMultipleTimeSeriesOperation; } });
var GetTimeSeriesOperation_1 = require("./Documents/Operations/TimeSeries/GetTimeSeriesOperation");
Object.defineProperty(exports, "GetTimeSeriesOperation", { enumerable: true, get: function () { return GetTimeSeriesOperation_1.GetTimeSeriesOperation; } });
var GetTimeSeriesStatisticsOperation_1 = require("./Documents/Operations/TimeSeries/GetTimeSeriesStatisticsOperation");
Object.defineProperty(exports, "GetTimeSeriesStatisticsOperation", { enumerable: true, get: function () { return GetTimeSeriesStatisticsOperation_1.GetTimeSeriesStatisticsOperation; } });
var RawTimeSeriesPolicy_1 = require("./Documents/Operations/TimeSeries/RawTimeSeriesPolicy");
Object.defineProperty(exports, "RawTimeSeriesPolicy", { enumerable: true, get: function () { return RawTimeSeriesPolicy_1.RawTimeSeriesPolicy; } });
var RemoveTimeSeriesPolicyOperation_1 = require("./Documents/Operations/TimeSeries/RemoveTimeSeriesPolicyOperation");
Object.defineProperty(exports, "RemoveTimeSeriesPolicyOperation", { enumerable: true, get: function () { return RemoveTimeSeriesPolicyOperation_1.RemoveTimeSeriesPolicyOperation; } });
var TimeSeriesBatchOperation_1 = require("./Documents/Operations/TimeSeries/TimeSeriesBatchOperation");
Object.defineProperty(exports, "TimeSeriesBatchOperation", { enumerable: true, get: function () { return TimeSeriesBatchOperation_1.TimeSeriesBatchOperation; } });
var TimeSeriesCollectionConfiguration_1 = require("./Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration");
Object.defineProperty(exports, "TimeSeriesCollectionConfiguration", { enumerable: true, get: function () { return TimeSeriesCollectionConfiguration_1.TimeSeriesCollectionConfiguration; } });
var TimeSeriesConfiguration_1 = require("./Documents/Operations/TimeSeries/TimeSeriesConfiguration");
Object.defineProperty(exports, "TimeSeriesConfiguration", { enumerable: true, get: function () { return TimeSeriesConfiguration_1.TimeSeriesConfiguration; } });
var TimeSeriesDetails_1 = require("./Documents/Operations/TimeSeries/TimeSeriesDetails");
Object.defineProperty(exports, "TimeSeriesDetails", { enumerable: true, get: function () { return TimeSeriesDetails_1.TimeSeriesDetails; } });
var TimeSeriesOperation_1 = require("./Documents/Operations/TimeSeries/TimeSeriesOperation");
Object.defineProperty(exports, "TimeSeriesOperation", { enumerable: true, get: function () { return TimeSeriesOperation_1.TimeSeriesOperation; } });
var TimeSeriesPolicy_1 = require("./Documents/Operations/TimeSeries/TimeSeriesPolicy");
Object.defineProperty(exports, "TimeSeriesPolicy", { enumerable: true, get: function () { return TimeSeriesPolicy_1.TimeSeriesPolicy; } });
var TimeSeriesRangeResult_1 = require("./Documents/Operations/TimeSeries/TimeSeriesRangeResult");
Object.defineProperty(exports, "TimeSeriesRangeResult", { enumerable: true, get: function () { return TimeSeriesRangeResult_1.TimeSeriesRangeResult; } });
__exportStar(require("./Auth/AuthOptions"), exports);
__exportStar(require("./Types/Callbacks"), exports);
__exportStar(require("./Types/Contracts"), exports);
__exportStar(require("./Types"), exports);
__exportStar(require("./Documents/Queries/IndexQuery"), exports);
__exportStar(require("./Documents/Queries/GroupBy"), exports);
__exportStar(require("./Documents/Queries/QueryOperator"), exports);
__exportStar(require("./Documents/Queries/SearchOperator"), exports);
__exportStar(require("./Documents/Queries/IIndexQuery"), exports);
__exportStar(require("./Documents/Queries/GroupByMethod"), exports);
__exportStar(require("./Documents/Queries/ProjectionBehavior"), exports);
__exportStar(require("./Documents/Queries/Spatial/SpatialCriteriaFactory"), exports);
__exportStar(require("./Documents/Queries/Spatial/SpatialCriteria"), exports);
__exportStar(require("./Documents/Queries/Spatial/CircleCriteria"), exports);
__exportStar(require("./Documents/Queries/Spatial/DynamicSpatialField"), exports);
__exportStar(require("./Documents/Queries/Spatial/WktCriteria"), exports);
__exportStar(require("./Documents/Queries/Spatial/PointField"), exports);
__exportStar(require("./Documents/Queries/Spatial/WktField"), exports);
__exportStar(require("./Documents/Queries/Facets/RangeBuilder"), exports);
__exportStar(require("./Documents/Queries/Facets/FacetBuilder"), exports);
__exportStar(require("./Documents/Queries/Facets/FacetAggregationField"), exports);
__exportStar(require("./Documents/Queries/Facets/Facet"), exports);
__exportStar(require("./Documents/Queries/Facets/RangeFacet"), exports);
__exportStar(require("./Documents/Queries/Facets/FacetBase"), exports);
__exportStar(require("./Documents/Queries/Facets/FacetSetup"), exports);
__exportStar(require("./Documents/Queries/Facets"), exports);
__exportStar(require("./Documents/Queries/Facets/AggregationRawDocumentQuery"), exports);
__exportStar(require("./Documents/Queries/QueryData"), exports);
__exportStar(require("./Documents/Queries/QueryOperationOptions"), exports);
__exportStar(require("./Documents/Queries/QueryResult"), exports);
__exportStar(require("./Documents/Queries/Highlighting/HighlightingOptions"), exports);
__exportStar(require("./Documents/Queries/Highlighting/HighlightingParameters"), exports);
__exportStar(require("./Documents/Queries/Highlighting/Hightlightings"), exports);
__exportStar(require("./Documents/Queries/Timings/QueryTimings"), exports);
__exportStar(require("./Documents/Queries/Facets/AggregationDocumentQuery"), exports);
__exportStar(require("./Documents/Queries/Facets/AggregationQueryBase"), exports);
__exportStar(require("./Documents/Queries/Facets/GenericRangeFacet"), exports);
__exportStar(require("./Documents/Queries/Facets/IAggregationDocumentQuery"), exports);
__exportStar(require("./Documents/Queries/Facets/IFacetBuilder"), exports);
__exportStar(require("./Documents/Queries/Facets/IFacetOperations"), exports);
__exportStar(require("./Documents/Queries/Explanation/ExplanationOptions"), exports);
__exportStar(require("./Documents/Queries/Explanation/Explanations"), exports);
__exportStar(require("./Documents/Queries/Highlighting/QueryHighlightings"), exports);
__exportStar(require("./Documents/Queries/Sorting/SorterDefinition"), exports);
__exportStar(require("./Documents/Queries/TimeSeries/ITimeSeriesQueryBuilder"), exports);
__exportStar(require("./Documents/Queries/TimeSeries/TimeSeriesAggregationResult"), exports);
__exportStar(require("./Documents/Queries/TimeSeries/TimeSeriesQueryBuilder"), exports);
__exportStar(require("./Documents/Queries/TimeSeries/TimeSeriesQueryResult"), exports);
__exportStar(require("./Documents/Queries/TimeSeries/TimeSeriesRangeAggregation"), exports);
__exportStar(require("./Documents/Queries/TimeSeries/TimeSeriesRawResult"), exports);
__exportStar(require("./Documents/Queries/TimeSeries/TypedTimeSeriesAggregationResult"), exports);
__exportStar(require("./Documents/Queries/TimeSeries/TypedTimeSeriesRangeAggregation"), exports);
__exportStar(require("./Documents/Queries/TimeSeries/TypedTimeSeriesRawResult"), exports);
__exportStar(require("./Documents/Queries/MoreLikeThis/IMoreLikeThisBuilderBase"), exports);
__exportStar(require("./Documents/Queries/MoreLikeThis/IMoreLikeThisOperations"), exports);
__exportStar(require("./Documents/Queries/MoreLikeThis/MoreLikeThisBase"), exports);
__exportStar(require("./Documents/Queries/MoreLikeThis/MoreLikeThisBuilder"), exports);
__exportStar(require("./Documents/Queries/MoreLikeThis/MoreLikeThisOptions"), exports);
__exportStar(require("./Documents/Queries/MoreLikeThis/MoreLikeThisStopWords"), exports);
__exportStar(require("./Documents/Queries/Suggestions/ISuggestionBuilder"), exports);
__exportStar(require("./Documents/Queries/Suggestions/ISuggestionDocumentQuery"), exports);
__exportStar(require("./Documents/Queries/Suggestions/ISuggestionOperations"), exports);
__exportStar(require("./Documents/Queries/Suggestions/StringDistanceTypes"), exports);
__exportStar(require("./Documents/Queries/Suggestions/SuggestionBuilder"), exports);
__exportStar(require("./Documents/Queries/Suggestions/SuggestionDocumentQuery"), exports);
__exportStar(require("./Documents/Queries/Suggestions/SuggestionOptions"), exports);
__exportStar(require("./Documents/Queries/Suggestions/SuggestionBase"), exports);
__exportStar(require("./Documents/Queries/Suggestions/SuggestionResult"), exports);
__exportStar(require("./Documents/Queries/Suggestions/SuggestionSortMode"), exports);
__exportStar(require("./Documents/Attachments"), exports);
__exportStar(require("./Documents/Operations/Attachments/GetAttachmentOperation"), exports);
__exportStar(require("./Documents/Operations/Attachments/AttachmentRequest"), exports);
__exportStar(require("./Documents/Operations/Analyzers/DeleteAnalyzerOperation"), exports);
__exportStar(require("./Documents/Operations/Analyzers/PutAnalyzersOperation"), exports);
__exportStar(require("./Documents/Changes/IndexChange"), exports);
__exportStar(require("./Documents/Changes/DatabaseChangesOptions"), exports);
__exportStar(require("./Documents/Changes/DocumentChange"), exports);
__exportStar(require("./Documents/Changes/TimeSeriesChange"), exports);
__exportStar(require("./Documents/Changes/CounterChange"), exports);
__exportStar(require("./Documents/Changes/IDatabaseChanges"), exports);
__exportStar(require("./Documents/Changes/DatabaseChange"), exports);
__exportStar(require("./Documents/Changes/OperationStatusChange"), exports);
__exportStar(require("./Documents/Changes/IDatabaseChanges"), exports);
__exportStar(require("./Documents/Changes/DatabaseChanges"), exports);
__exportStar(require("./Documents/Changes/IConnectableChanges"), exports);
__exportStar(require("./Documents/Changes/IChangesObservable"), exports);
__exportStar(require("./Documents/Changes/ChangesObservable"), exports);
__exportStar(require("./Documents/Changes/DatabaseConnectionState"), exports);
__exportStar(require("./Documents/Changes/IChangesConnectionState"), exports);
__exportStar(require("./Documents/Identity/HiloIdGenerator"), exports);
__exportStar(require("./Documents/Identity/MultiDatabaseHiLoIdGenerator"), exports);
__exportStar(require("./Documents/Identity/MultiTypeHiLoIdGenerator"), exports);
__exportStar(require("./Documents/Identity/HiloRangeValue"), exports);
__exportStar(require("./Documents/Identity/MultiDatabaseHiLoIdGenerator"), exports);
__exportStar(require("./Documents/Smuggler/DatabaseItemType"), exports);
__exportStar(require("./Documents/Smuggler/DatabaseRecordItemType"), exports);
__exportStar(require("./Documents/Smuggler/DatabaseSmuggler"), exports);
__exportStar(require("./Documents/Smuggler/DatabaseSmugglerExportOptions"), exports);
__exportStar(require("./Documents/Smuggler/IDatabaseSmugglerExportOptions"), exports);
__exportStar(require("./Documents/Smuggler/DatabaseSmugglerImportOptions"), exports);
__exportStar(require("./Documents/Smuggler/IDatabaseSmugglerImportOptions"), exports);
__exportStar(require("./Documents/Smuggler/DatabaseSmugglerOptions"), exports);
__exportStar(require("./Documents/Smuggler/IDatabaseSmugglerOptions"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/CertificateDefinition"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/CertificateRawData"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/CreateClientCertificateOperation"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/DatabaseAccess"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/DeleteCertificateOperation"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/GetCertificateOperation"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/GetCertificatesOperation"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/GetCertificatesResponse"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/PutClientCertificateOperation"), exports);
__exportStar(require("./ServerWide/Operations/Certificates/SecurityClearance"), exports);
__exportStar(require("./ServerWide/Operations/AddDatabaseNodeOperation"), exports);
__exportStar(require("./ServerWide/Operations/PromoteDatabaseNodeOperation"), exports);
__exportStar(require("./ServerWide/Operations/Analyzers/DeleteServerWideAnalyzerOperation"), exports);
__exportStar(require("./ServerWide/Operations/Analyzers/PutServerWideAnalyzersOperation"), exports);
__exportStar(require("./ServerWide/Operations/DocumentsCompression/DocumentCompressionConfigurationResult"), exports);
__exportStar(require("./ServerWide/Operations/DocumentsCompression/UpdateDocumentsCompressionConfigurationOperation"), exports);
__exportStar(require("./ServerWide/Operations/OngoingTasks/IServerWideTask"), exports);
__exportStar(require("./ServerWide/Operations/OngoingTasks/DeleteServerWideTaskOperation"), exports);
__exportStar(require("./ServerWide/Operations/OngoingTasks/SetDatabasesLockOperation"), exports);
__exportStar(require("./ServerWide/Operations/OngoingTasks/ToggleServerWideTaskStateOperation"), exports);
__exportStar(require("./ServerWide/Operations/OngoingTasks/GetServerWideExternalReplicationOperation"), exports);
__exportStar(require("./ServerWide/Operations/OngoingTasks/PutServerWideExternalReplicationOperation"), exports);
__exportStar(require("./ServerWide/Operations/OngoingTasks/ServerWideTaskResponse"), exports);
__exportStar(require("./ServerWide/Operations/OngoingTasks/ServerWideExternalReplication"), exports);
__exportStar(require("./ServerWide/Operations/Sorters/DeleteServerWideSorterOperation"), exports);
__exportStar(require("./ServerWide/Operations/Sorters/PutServerWideSortersOperation"), exports);
var ObjectMapper_1 = require("./Mapping/ObjectMapper");
Object.defineProperty(exports, "TypesAwareObjectMapper", { enumerable: true, get: function () { return ObjectMapper_1.TypesAwareObjectMapper; } });
var Mapping_1 = require("./Mapping");
Object.defineProperty(exports, "Mapping", { enumerable: true, get: function () { return Mapping_1.Mapping; } });
const Json = require("./Mapping/Json");
exports.Json = Json;
var Constants_1 = require("./Constants");
Object.defineProperty(exports, "CONSTANTS", { enumerable: true, get: function () { return Constants_1.CONSTANTS; } });
const DocumentStore_1 = require("./Documents/DocumentStore");
exports.default = DocumentStore_1.DocumentStore;
