// ravendb/dist/ServerWide/Operations/Sorters/PutServerWideSortersOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PutServerWideSortersOperation = void 0;
const Exceptions_1 = require("../../../Exceptions");
const RavenCommand_1 = require("../../../Http/RavenCommand");
const RaftIdGenerator_1 = require("../../../Utility/RaftIdGenerator");
class PutServerWideSortersOperation {
    constructor(...sortersToAdd) {
        if (!sortersToAdd || sortersToAdd.length === 0) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "SortersToAdd cannot be null or empty");
        }
        this._sortersToAdd = sortersToAdd;
    }
    getCommand(conventions) {
        return new PutServerWideSortersCommand(conventions, this._sortersToAdd);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.PutServerWideSortersOperation = PutServerWideSortersOperation;
class PutServerWideSortersCommand extends RavenCommand_1.RavenCommand {
    constructor(conventions, sortersToAdd) {
        super();
        if (!conventions) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Conventions cannot be null");
        }
        if (!sortersToAdd) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "SortersToAdd cannot be null");
        }
        this._sortersToAdd = sortersToAdd.map(x => {
            if (!x.name) {
                (0, Exceptions_1.throwError)("InvalidArgumentException", "Sorter name cannot be null");
            }
            return conventions.objectMapper.toObjectLiteral(x);
        });
    }
    createRequest(node) {
        const uri = node.url + "/admin/sorters";
        const headers = this._headers()
            .typeAppJson().build();
        const body = this._serializer.serialize({
            Sorters: this._sortersToAdd
        });
        return {
            uri,
            method: "PUT",
            headers,
            body
        };
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_1.RaftIdGenerator.newId();
    }
}
