// ravendb/dist/ServerWide/Operations/Certificates/DeleteCertificateOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteCertificateOperation = void 0;
const Exceptions_1 = require("../../../Exceptions");
const RavenCommand_1 = require("../../../Http/RavenCommand");
const RaftIdGenerator_1 = require("../../../Utility/RaftIdGenerator");
class DeleteCertificateOperation {
    constructor(thumbprint) {
        if (!thumbprint) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Thumbprint cannot be null.");
        }
        this._thumbprint = thumbprint;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new DeleteCertificateCommand(this._thumbprint);
    }
}
exports.DeleteCertificateOperation = DeleteCertificateOperation;
class DeleteCertificateCommand extends RavenCommand_1.RavenCommand {
    constructor(thumbprint) {
        super();
        if (!thumbprint) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Thumbprint cannot be null");
        }
        this._thumbprint = thumbprint;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/admin/certificates?thumbprint=" + encodeURIComponent(this._thumbprint);
        return {
            uri,
            method: "DELETE"
        };
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_1.RaftIdGenerator.newId();
    }
}
