// ravendb/dist/Documents/Operations/Identities/GetIdentitiesOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetIdentitiesCommand = exports.GetIdentitiesOperation = void 0;
const RavenCommand_1 = require("../../../Http/RavenCommand");
class GetIdentitiesOperation {
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new GetIdentitiesCommand();
    }
}
exports.GetIdentitiesOperation = GetIdentitiesOperation;
class GetIdentitiesCommand extends RavenCommand_1.RavenCommand {
    constructor() {
        super();
    }
    get isReadRequest() {
        return true;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/debug/identities";
        return { uri };
    }
    async setResponseAsync(bodyStream, fromCache) {
        let body = null;
        this.result = await this._pipeline()
            .parseJsonSync()
            .collectBody(b => body = b)
            .process(bodyStream);
        return body;
    }
}
exports.GetIdentitiesCommand = GetIdentitiesCommand;
