// ravendb/dist/Documents/Operations/Indexes/DeleteIndexErrorsOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteIndexErrorsOperation = void 0;
const RavenCommand_1 = require("../../../Http/RavenCommand");
class DeleteIndexErrorsOperation {
    constructor(indexNames) {
        this._indexNames = indexNames;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new DeleteIndexErrorsCommand(this._indexNames);
    }
}
exports.DeleteIndexErrorsOperation = DeleteIndexErrorsOperation;
class DeleteIndexErrorsCommand extends RavenCommand_1.RavenCommand {
    constructor(indexNames) {
        super();
        this._indexNames = indexNames;
    }
    createRequest(node) {
        let uri = node.url + "/databases/" + node.database + "/indexes/errors";
        if (this._indexNames && this._indexNames.length) {
            uri += "?";
            for (const indexName of this._indexNames) {
                uri += "&name=" + this._urlEncode(indexName);
            }
        }
        return {
            uri,
            method: "DELETE"
        };
    }
    get isReadRequest() {
        return false;
    }
}
