// ravendb/dist/Documents/Operations/Indexes/ResetIndexOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResetIndexCommand = exports.ResetIndexOperation = void 0;
const Exceptions_1 = require("../../../Exceptions");
const RavenCommand_1 = require("../../../Http/RavenCommand");
class ResetIndexOperation {
    constructor(indexName) {
        if (!indexName) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Index name cannot be null.");
        }
        this._indexName = indexName;
    }
    getCommand(conventions) {
        return new ResetIndexCommand(this._indexName);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.ResetIndexOperation = ResetIndexOperation;
class ResetIndexCommand extends RavenCommand_1.RavenCommand {
    constructor(indexName) {
        super();
        if (!indexName) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Index name cannot be null.");
        }
        this._indexName = indexName;
        this._responseType = "Empty";
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/indexes?name=" + encodeURIComponent(this._indexName);
        return { method: "RESET", uri };
    }
    get isReadRequest() {
        return false;
    }
}
exports.ResetIndexCommand = ResetIndexCommand;
