// ravendb/dist/Documents/Queries/Spatial/SpatialCriteriaFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpatialCriteriaFactory = void 0;
const Constants_1 = require("../../../Constants");
const WktCriteria_1 = require("./WktCriteria");
const CircleCriteria_1 = require("./CircleCriteria");
const TypeUtil_1 = require("../../../Utility/TypeUtil");
class SpatialCriteriaFactory {
    constructor() {
    }
    relatesToShape(shapeWkt, relation, units, distErrorPercent) {
        if (TypeUtil_1.TypeUtil.isNullOrUndefined(distErrorPercent)) {
            distErrorPercent = Constants_1.CONSTANTS.Documents.Indexing.Spatial.DEFAULT_DISTANCE_ERROR_PCT;
        }
        return new WktCriteria_1.WktCriteria(shapeWkt, relation, units, distErrorPercent);
    }
    static _normalizeArgs(distErrorPercentOrUnits, distErrorPercent) {
        let units = null;
        if (TypeUtil_1.TypeUtil.isString(distErrorPercentOrUnits)) {
            units = distErrorPercentOrUnits;
        }
        else if (TypeUtil_1.TypeUtil.isNumber(distErrorPercentOrUnits)) {
            distErrorPercent = distErrorPercentOrUnits;
            units = null;
        }
        distErrorPercent = distErrorPercent ?? Constants_1.CONSTANTS.Documents.Indexing.Spatial.DEFAULT_DISTANCE_ERROR_PCT;
        return { units, distErrorPercent };
    }
    intersects(shapeWkt, distErrorPercentOrUnits, distErrorPercent) {
        const args = SpatialCriteriaFactory._normalizeArgs(distErrorPercentOrUnits, distErrorPercent);
        return this.relatesToShape(shapeWkt, "Intersects", args.units, args.distErrorPercent);
    }
    contains(shapeWkt, distErrorPercentOrUnits, distErrorPercent) {
        const args = SpatialCriteriaFactory._normalizeArgs(distErrorPercentOrUnits, distErrorPercent);
        return this.relatesToShape(shapeWkt, "Contains", args.units, args.distErrorPercent);
    }
    disjoint(shapeWkt, distErrorPercentOrUnits, distErrorPercent) {
        const args = SpatialCriteriaFactory._normalizeArgs(distErrorPercentOrUnits, distErrorPercent);
        return this.relatesToShape(shapeWkt, "Disjoint", args.units, args.distErrorPercent);
    }
    within(shapeWkt, distErrorPercentOrUnits, distErrorPercent) {
        const args = SpatialCriteriaFactory._normalizeArgs(distErrorPercentOrUnits, distErrorPercent);
        return this.relatesToShape(shapeWkt, "Within", args.units, args.distErrorPercent);
    }
    withinRadius(radius, latitude, longitude, radiusUnits = null, distErrorPercent) {
        distErrorPercent = distErrorPercent ?? Constants_1.CONSTANTS.Documents.Indexing.Spatial.DEFAULT_DISTANCE_ERROR_PCT;
        return new CircleCriteria_1.CircleCriteria(radius, latitude, longitude, radiusUnits, "Within", distErrorPercent);
    }
}
exports.SpatialCriteriaFactory = SpatialCriteriaFactory;
SpatialCriteriaFactory.INSTANCE = new SpatialCriteriaFactory();
