// ravendb/dist/Documents/Operations/Etl/AddEtlOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddEtlOperation = void 0;
const RavenCommand_1 = require("../../../Http/RavenCommand");
const RaftIdGenerator_1 = require("../../../Utility/RaftIdGenerator");
class AddEtlOperation {
    constructor(configuration) {
        this._configuration = configuration;
    }
    getCommand(conventions) {
        return new AddEtlCommand(conventions, this._configuration);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.AddEtlOperation = AddEtlOperation;
class AddEtlCommand extends RavenCommand_1.RavenCommand {
    constructor(conventions, configuration) {
        super();
        this._conventions = conventions;
        this._configuration = configuration;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/etl";
        const body = JSON.stringify(this._configuration.serialize(this._conventions));
        const headers = this._headers().typeAppJson().build();
        return {
            uri,
            method: "PUT",
            body,
            headers
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_1.RaftIdGenerator.newId();
    }
}
