// ravendb/dist/Documents/Commands/MultiGet/GetResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetResponse = void 0;
const StatusCode_1 = require("../../../Http/StatusCode");
class GetResponse {
    constructor() {
        this._headers = {};
    }
    static create(data) {
        return Object.assign(new GetResponse(), data);
    }
    get elapsed() {
        return this._elapsed;
    }
    set elapsed(elapsed) {
        this._elapsed = elapsed;
    }
    get result() {
        return this._result;
    }
    set result(result) {
        this._result = result;
    }
    get headers() {
        return this._headers;
    }
    set headers(headers) {
        this._headers = headers;
    }
    get statusCode() {
        return this._statusCode;
    }
    set statusCode(statusCode) {
        this._statusCode = statusCode;
    }
    get forceRetry() {
        return this._forceRetry;
    }
    set forceRetry(forceRetry) {
        this._forceRetry = forceRetry;
    }
    requestHasErrors() {
        switch (this._statusCode) {
            case 0:
            case StatusCode_1.StatusCodes.Ok:
            case StatusCode_1.StatusCodes.Created:
            case StatusCode_1.StatusCodes.NonAuthoritativeInformation:
            case StatusCode_1.StatusCodes.NoContent:
            case StatusCode_1.StatusCodes.NotModified:
            case StatusCode_1.StatusCodes.NotFound:
                return false;
            default:
                return true;
        }
    }
}
exports.GetResponse = GetResponse;
