// ravendb/dist/ServerWide/Operations/Analyzers/DeleteServerWideAnalyzerOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteServerWideAnalyzerOperation = void 0;
const Exceptions_1 = require("../../../Exceptions");
const RavenCommand_1 = require("../../../Http/RavenCommand");
const RaftIdGenerator_1 = require("../../../Utility/RaftIdGenerator");
class DeleteServerWideAnalyzerOperation {
    constructor(analyzerName) {
        if (!analyzerName) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "AnalyzerName cannot be null");
        }
        this._analyzerName = analyzerName;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new DeleteServerWideAnalyzerCommand(this._analyzerName);
    }
}
exports.DeleteServerWideAnalyzerOperation = DeleteServerWideAnalyzerOperation;
class DeleteServerWideAnalyzerCommand extends RavenCommand_1.RavenCommand {
    constructor(analyzerName) {
        super();
        if (!analyzerName) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "AnalyzerName cannot be null");
        }
        this._analyzerName = analyzerName;
    }
    createRequest(node) {
        const uri = node.url + "/admin/analyzers?name=" + this._urlEncode(this._analyzerName);
        return {
            uri,
            method: "DELETE"
        };
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_1.RaftIdGenerator.newId();
    }
}
