// ravendb/dist/Documents/Operations/Counters/CounterOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CounterOperation = void 0;
class CounterOperation {
    serialize() {
        return {
            Type: this._type,
            CounterName: this._counterName,
            Delta: this._delta
        };
    }
    get type() {
        return this._type;
    }
    set type(value) {
        this._type = value;
    }
    get counterName() {
        return this._counterName;
    }
    set counterName(value) {
        this._counterName = value;
    }
    get delta() {
        return this._delta;
    }
    set delta(value) {
        this._delta = value;
    }
    get changeVector() {
        return this._changeVector;
    }
    set changeVector(changeVector) {
        this._changeVector = changeVector;
    }
    static create(counterName, type, delta) {
        const operation = new CounterOperation();
        operation.counterName = counterName;
        operation.type = type;
        if (arguments.length === 3) {
            operation.delta = delta;
        }
        return operation;
    }
}
exports.CounterOperation = CounterOperation;
