// ravendb/dist/Documents/Operations/Counters/GetCountersOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetCounterValuesCommand = exports.GetCountersOperation = void 0;
const RavenCommand_1 = require("../../../Http/RavenCommand");
const Exceptions_1 = require("../../../Exceptions");
const DocumentCountersOperation_1 = require("./DocumentCountersOperation");
const CounterOperation_1 = require("./CounterOperation");
const CounterBatch_1 = require("./CounterBatch");
const StringBuilder_1 = require("../../../Utility/StringBuilder");
class GetCountersOperation {
    constructor(docId, counters, returnFullResults = false) {
        this._docId = docId;
        if (counters) {
            this._counters = Array.isArray(counters) ? counters : [counters];
        }
        else {
            this._counters = [];
        }
        this._returnFullResults = returnFullResults;
    }
    getCommand(store, conventions, cache) {
        return new GetCounterValuesCommand(this._docId, this._counters, this._returnFullResults, conventions);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.GetCountersOperation = GetCountersOperation;
class GetCounterValuesCommand extends RavenCommand_1.RavenCommand {
    constructor(docId, counters, returnFullResults, conventions) {
        super();
        if (!docId) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "DocId cannot be null");
        }
        this._docId = docId;
        this._counters = counters;
        this._returnFullResults = returnFullResults;
        this._conventions = conventions;
    }
    createRequest(node) {
        const pathBuilder = new StringBuilder_1.StringBuilder(node.url);
        pathBuilder.append("/databases/")
            .append(node.database)
            .append("/counters?docId=")
            .append(encodeURIComponent(this._docId));
        let req = { uri: null, method: "GET" };
        if (this._counters.length > 0) {
            if (this._counters && this._counters.length > 1) {
                req = this._prepareRequestWithMultipleCounters(pathBuilder, req);
            }
            else {
                pathBuilder.append("&counter=")
                    .append(encodeURIComponent(this._counters[0]));
            }
        }
        if (this._returnFullResults && req.method === "GET") {
            pathBuilder.append("&full=true");
        }
        req.uri = pathBuilder.toString();
        return req;
    }
    _prepareRequestWithMultipleCounters(pathBuilder, request) {
        const [uniqueNames, sumLength] = this._getOrderedUniqueNames();
        if (sumLength < 1024) {
            for (const uniqueName of uniqueNames) {
                pathBuilder
                    .append("&counter=")
                    .append(encodeURIComponent(uniqueName || ""));
            }
        }
        else {
            request = { method: "POST" };
            const docOps = new DocumentCountersOperation_1.DocumentCountersOperation();
            docOps.documentId = this._docId;
            docOps.operations = [];
            for (const counter of uniqueNames) {
                const counterOperation = new CounterOperation_1.CounterOperation();
                counterOperation.type = "Get";
                counterOperation.counterName = counter;
                docOps.operations.push(counterOperation);
            }
            const batch = new CounterBatch_1.CounterBatch();
            batch.documents = [docOps];
            batch.replyWithAllNodesValues = this._returnFullResults;
            request.body = JSON.stringify(batch.serialize());
            request.headers = this._headers().typeAppJson().build();
        }
        return request;
    }
    _getOrderedUniqueNames() {
        const uniqueNames = new Set();
        const orderedUniqueNames = [];
        let sum = 0;
        for (const counter of this._counters) {
            const containsCounter = uniqueNames.has(counter);
            if (!containsCounter) {
                uniqueNames.add(counter);
                orderedUniqueNames.push(counter);
                sum += counter?.length || 0;
            }
        }
        return [orderedUniqueNames, sum];
    }
    get isReadRequest() {
        return true;
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            return;
        }
        let body = "";
        this.result = await this._defaultPipeline(_ => body += _).process(bodyStream);
        return body;
    }
}
exports.GetCounterValuesCommand = GetCounterValuesCommand;
