// ravendb/dist/Documents/Operations/TransactionsRecording/StartTransactionsRecordingOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StartTransactionsRecordingOperation = void 0;
const RavenCommand_1 = require("../../../Http/RavenCommand");
const Exceptions_1 = require("../../../Exceptions");
class StartTransactionsRecordingOperation {
    constructor(filePath) {
        if (!filePath) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "FilePath cannot be null");
        }
        this._filePath = filePath;
    }
    getCommand(conventions) {
        return new StartTransactionsRecordingCommand(this._filePath);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.StartTransactionsRecordingOperation = StartTransactionsRecordingOperation;
class StartTransactionsRecordingCommand extends RavenCommand_1.RavenCommand {
    constructor(filePath) {
        super();
        this._filePath = filePath;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/transactions/start-recording";
        const body = this._serializer.serialize({
            File: this._filePath
        });
        return {
            uri,
            method: "POST",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    get isReadRequest() {
        return false;
    }
    async setResponseAsync(bodyStream, fromCache) {
        return this._parseResponseDefaultAsync(bodyStream);
    }
}
