// ravendb/dist/Documents/Indexes/Counters/AbstractGenericCountersIndexCreationTask.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractGenericCountersIndexCreationTask = void 0;
const AbstractIndexCreationTaskBase_1 = require("../AbstractIndexCreationTaskBase");
const Spatial_1 = require("../Spatial");
const Constants_1 = require("../../../Constants");
class AbstractGenericCountersIndexCreationTask extends AbstractIndexCreationTaskBase_1.AbstractIndexCreationTaskBase {
    constructor() {
        super();
        this._storesStrings = {};
        this._indexesStrings = {};
        this._analyzersStrings = {};
        this._indexSuggestions = new Set();
        this._termVectorsStrings = {};
        this._spatialOptionsStrings = {};
    }
    get isMapReduce() {
        return !!this._reduce;
    }
    index(field, indexing) {
        this._indexesStrings[field] = indexing;
    }
    spatial(field, indexing) {
        this._spatialOptionsStrings[field] = indexing(new Spatial_1.SpatialOptionsFactory());
    }
    storeAllFields(storage) {
        this._storesStrings[Constants_1.CONSTANTS.Documents.Indexing.Fields.ALL_FIELDS] = storage;
    }
    store(field, storage) {
        this._storesStrings[field] = storage;
    }
    analyze(field, analyzer) {
        this._analyzersStrings[field] = analyzer;
    }
    termVector(field, termVector) {
        this._termVectorsStrings[field] = termVector;
    }
    suggestion(field) {
        this._indexSuggestions.add(field);
    }
}
exports.AbstractGenericCountersIndexCreationTask = AbstractGenericCountersIndexCreationTask;
