// ravendb/dist/Documents/Indexes/Counters/AbstractRawJavaScriptCountersIndexCreationTask.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractRawJavaScriptCountersIndexCreationTask = void 0;
const CountersIndexDefinition_1 = require("./CountersIndexDefinition");
const AbstractIndexCreationTaskBase_1 = require("../AbstractIndexCreationTaskBase");
class AbstractRawJavaScriptCountersIndexCreationTask extends AbstractIndexCreationTaskBase_1.AbstractIndexCreationTaskBase {
    constructor() {
        super();
        this._definition = new CountersIndexDefinition_1.CountersIndexDefinition();
    }
    get maps() {
        return this._definition.maps;
    }
    set maps(value) {
        this._definition.maps = value;
    }
    get fields() {
        return this._definition.fields;
    }
    set fields(value) {
        this._definition.fields = value;
    }
    get reduce() {
        return this._definition.reduce;
    }
    set reduce(value) {
        this._definition.reduce = value;
    }
    get isMapReduce() {
        return !!this.reduce;
    }
    get outputReduceToCollection() {
        return this._definition.outputReduceToCollection;
    }
    set outputReduceToCollection(value) {
        this._definition.outputReduceToCollection = value;
    }
    get patternReferencesCollectionName() {
        return this._definition.patternReferencesCollectionName;
    }
    set patternReferencesCollectionName(value) {
        this._definition.patternReferencesCollectionName = value;
    }
    get patternForOutputReduceToCollectionReferences() {
        return this._definition.patternForOutputReduceToCollectionReferences;
    }
    set patternForOutputReduceToCollectionReferences(value) {
        this._definition.patternForOutputReduceToCollectionReferences = value;
    }
    createIndexDefinition() {
        this._definition.name = this.getIndexName();
        this._definition.type = this.isMapReduce ? "JavaScriptMapReduce" : "JavaScriptMap";
        this._definition.name = this.getIndexName();
        if (this.additionalSources) {
            this._definition.additionalSources = this.additionalSources;
        }
        else {
            this._definition.additionalSources = {};
        }
        this._definition.additionalAssemblies = this.additionalAssemblies || [];
        this._definition.configuration = this.configuration;
        this._definition.lockMode = this.lockMode;
        this._definition.priority = this.priority;
        this._definition.state = this.state;
        this._definition.deploymentMode = this.deploymentMode;
        return this._definition;
    }
}
exports.AbstractRawJavaScriptCountersIndexCreationTask = AbstractRawJavaScriptCountersIndexCreationTask;
