// ravendb/dist/ServerWide/Operations/DocumentsCompression/UpdateDocumentsCompressionConfigurationOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateDocumentsCompressionConfigurationOperation = void 0;
const Exceptions_1 = require("../../../Exceptions");
const RavenCommand_1 = require("../../../Http/RavenCommand");
const RaftIdGenerator_1 = require("../../../Utility/RaftIdGenerator");
class UpdateDocumentsCompressionConfigurationOperation {
    constructor(configuration) {
        if (!configuration) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Configuration cannot be null");
        }
        this._documentsCompressionConfiguration = configuration;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new UpdateDocumentCompressionConfigurationCommand(this._documentsCompressionConfiguration);
    }
}
exports.UpdateDocumentsCompressionConfigurationOperation = UpdateDocumentsCompressionConfigurationOperation;
class UpdateDocumentCompressionConfigurationCommand extends RavenCommand_1.RavenCommand {
    constructor(configuration) {
        super();
        if (!configuration) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Configuration cannot be null");
        }
        this._documentsCompressionConfiguration = configuration;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/documents-compression/config";
        const headers = this._headers()
            .typeAppJson().build();
        const body = this._serializer.serialize(this._documentsCompressionConfiguration);
        return {
            uri,
            method: "POST",
            headers,
            body
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            return;
        }
        return await this._parseResponseDefaultAsync(bodyStream);
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_1.RaftIdGenerator.newId();
    }
}
