// ravendb/dist/ServerWide/Operations/Configuration/DeleteServerWideBackupConfigurationOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteServerWideBackupConfigurationOperation = void 0;
const index_1 = require("../../../Exceptions/index");
const RavenCommand_1 = require("../../../Http/RavenCommand");
const RaftIdGenerator_1 = require("../../../Utility/RaftIdGenerator");
class DeleteServerWideBackupConfigurationOperation {
    constructor(name) {
        if (!name) {
            (0, index_1.throwError)("InvalidArgumentException", "Name cannot be null");
        }
        this._name = name;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new DeleteServerWideBackupConfigurationCommand(this._name);
    }
}
exports.DeleteServerWideBackupConfigurationOperation = DeleteServerWideBackupConfigurationOperation;
class DeleteServerWideBackupConfigurationCommand extends RavenCommand_1.RavenCommand {
    constructor(name) {
        super();
        if (!name) {
            (0, index_1.throwError)("InvalidArgumentException", "Name cannot be null");
        }
        this._responseType = "Empty";
        this._name = name;
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_1.RaftIdGenerator.newId();
    }
    createRequest(node) {
        const uri = node.url + "/admin/configuration/server-wide/backup?name=" + encodeURIComponent(this._name);
        return {
            method: "DELETE",
            uri
        };
    }
}
