// ravendb/dist/ServerWide/Operations/Configuration/GetDatabaseSettingsOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetDatabaseSettingsOperation = void 0;
const Exceptions_1 = require("../../../Exceptions");
const RavenCommand_1 = require("../../../Http/RavenCommand");
class GetDatabaseSettingsOperation {
    get resultType() {
        return "CommandResult";
    }
    constructor(databaseName) {
        if (!databaseName) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "DatabaseName cannot be null");
        }
        this._databaseName = databaseName;
    }
    getCommand(conventions) {
        return new GetDatabaseSettingsCommand(this._databaseName);
    }
}
exports.GetDatabaseSettingsOperation = GetDatabaseSettingsOperation;
class GetDatabaseSettingsCommand extends RavenCommand_1.RavenCommand {
    constructor(databaseName) {
        super();
        if (!databaseName) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "DatabaseName cannot be null");
        }
        this._databaseName = databaseName;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + this._databaseName + "/admin/record";
        return {
            method: "GET",
            uri
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        let body;
        const result = await this._pipeline()
            .parseJsonSync()
            .collectBody(_ => body = _)
            .process(bodyStream);
        this.result = {
            settings: result.Settings
        };
        return body;
    }
}
