// ravendb/dist/ServerWide/Operations/Configuration/GetServerWideClientConfigurationOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetServerWideClientConfigurationOperation = void 0;
const RavenCommand_1 = require("../../../Http/RavenCommand");
class GetServerWideClientConfigurationOperation {
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new GetServerWideClientConfigurationCommand();
    }
}
exports.GetServerWideClientConfigurationOperation = GetServerWideClientConfigurationOperation;
class GetServerWideClientConfigurationCommand extends RavenCommand_1.RavenCommand {
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/configuration/client";
        return {
            uri,
            method: "GET"
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
}
