// ravendb/dist/ServerWide/Operations/Configuration/PutServerWideBackupConfigurationOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PutServerWideBackupConfigurationOperation = void 0;
const index_1 = require("../../../Exceptions/index");
const RavenCommand_1 = require("../../../Http/RavenCommand");
const RaftIdGenerator_1 = require("../../../Utility/RaftIdGenerator");
class PutServerWideBackupConfigurationOperation {
    constructor(configuration) {
        if (!configuration) {
            (0, index_1.throwError)("InvalidArgumentException", "Configuration cannot be null");
        }
        this._configuration = configuration;
    }
    getCommand(conventions) {
        return new PutServerWideBackupConfigurationCommand(this._configuration);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.PutServerWideBackupConfigurationOperation = PutServerWideBackupConfigurationOperation;
class PutServerWideBackupConfigurationCommand extends RavenCommand_1.RavenCommand {
    constructor(configuration) {
        super();
        if (!configuration) {
            (0, index_1.throwError)("InvalidArgumentException", "Configuration cannot be null");
        }
        this._configuration = configuration;
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_1.RaftIdGenerator.newId();
    }
    createRequest(node) {
        const uri = node.url + "/admin/configuration/server-wide/backup";
        const body = this._serializer.serialize(this._configuration);
        return {
            uri,
            method: "PUT",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        return this._parseResponseDefaultAsync(bodyStream);
    }
}
