// ravendb/dist/ServerWide/Operations/Configuration/PutServerWideClientConfigurationOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PutServerWideClientConfigurationOperation = void 0;
const Exceptions_1 = require("../../../Exceptions");
const RavenCommand_1 = require("../../../Http/RavenCommand");
const RaftIdGenerator_1 = require("../../../Utility/RaftIdGenerator");
class PutServerWideClientConfigurationOperation {
    constructor(configuration) {
        if (!configuration) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Configuration cannot be null");
        }
        this._configuration = configuration;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new PutServerWideClientConfigurationCommand(conventions, this._configuration);
    }
}
exports.PutServerWideClientConfigurationOperation = PutServerWideClientConfigurationOperation;
class PutServerWideClientConfigurationCommand extends RavenCommand_1.RavenCommand {
    constructor(conventions, configuration) {
        super();
        if (!conventions) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Conventions cannot be null");
        }
        if (!configuration) {
            (0, Exceptions_1.throwError)("InvalidArgumentException", "Configuration cannot be null");
        }
        this._configuration = configuration;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/admin/configuration/client";
        const body = this._serializer.serialize(this._configuration);
        return {
            uri,
            method: "PUT",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_1.RaftIdGenerator.newId();
    }
}
