// ravendb/dist/Documents/Operations/Backups/StartBackupOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StartBackupOperation = void 0;
const RavenCommand_1 = require("../../../Http/RavenCommand");
class StartBackupOperation {
    constructor(isFullBackup, taskId) {
        this._isFullBackup = isFullBackup;
        this._taskId = taskId;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new StartBackupCommand(this._isFullBackup, this._taskId);
    }
}
exports.StartBackupOperation = StartBackupOperation;
class StartBackupCommand extends RavenCommand_1.RavenCommand {
    constructor(isFullBackup, taskId) {
        super();
        this._isFullBackup = isFullBackup;
        this._taskId = taskId;
    }
    get isReadRequest() {
        return true;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database
            + "/admin/backup/database?isFullBackup=" + (this._isFullBackup ? "true" : "false")
            + "&taskId=" + this._taskId;
        return {
            uri,
            method: "POST"
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
}
