// ravendb/dist/Mapping/Json/Conventions/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getObjectKeyCaseTransformProfile = exports.MULTI_GET_KEY_CASE_TRANSFORM_PROFILE = exports.DOCUMENT_LOAD_KEY_CASE_TRANSFORM_PROFILE = void 0;
const Constants_1 = require("../../../Constants");
exports.DOCUMENT_LOAD_KEY_CASE_TRANSFORM_PROFILE = {
    ignorePaths: [Constants_1.CONSTANTS.Documents.Metadata.IGNORE_CASE_TRANSFORM_REGEX],
    ignoreKeys: [/^@/],
    paths: [
        {
            transform: "camel",
            path: /@metadata\.@attachments/
        }
    ]
};
exports.MULTI_GET_KEY_CASE_TRANSFORM_PROFILE = {
    ignorePaths: [/^headers\./i]
};
function getObjectKeyCaseTransformProfile(defaultTransform, profile) {
    switch (profile) {
        case "DOCUMENT_LOAD":
        case "DOCUMENT_QUERY":
            return Object.assign({ defaultTransform }, exports.DOCUMENT_LOAD_KEY_CASE_TRANSFORM_PROFILE);
        default:
            return { defaultTransform };
    }
}
exports.getObjectKeyCaseTransformProfile = getObjectKeyCaseTransformProfile;
